{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pybedtools\n",
    "import scipy\n",
    "import pandas as pd\n",
    "import numpy as np # need for  stats\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import re\n",
    "from scipy.stats import mannwhitneyu\n",
    "import os\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load modules\n",
    "\n",
    "import glob\n",
    "from scipy.stats import wilcoxon\n",
    "from pybedtools import BedTool\n",
    "from scipy.stats import spearmanr\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#First we need to define the base dirs\n",
    "DIRS = {}\n",
    "DIRS['BASE2'] = '/home/anjuni/analysis'\n",
    "DIRS['FIGURES'] = os.path.join(DIRS['BASE2'], 'figures')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIRS['COVERAGE'] = os.path.join(DIRS['BASE2'], 'coverage')\n",
    "DIRS['GENE'] = os.path.join(DIRS['COVERAGE'], 'gene_level')\n",
    "DIRS['GENE_ANNO'] = os.path.join(DIRS['GENE'], 'gene_anno')\n",
    "DIRS['BOTH_U_D'] = os.path.join(DIRS['GENE'], 'both_upstream_downstream')\n",
    "DIRS['DOWN_STR'] = os.path.join(DIRS['GENE'], 'downstream')\n",
    "DIRS['UP_STR'] = os.path.join(DIRS['GENE'], 'upstream')\n",
    "DIRS['TSS'] = os.path.join(DIRS['GENE'], 'tss_6mA_only')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "DIRS['BASE1'] = '/home/anjuni/methylation_calling/pacbio'\n",
    "\n",
    "DIRS['BED_INPUT'] = os.path.join(DIRS['BASE2'], 'bedtools_output', 'sequencing_comparison')\n",
    "DIRS['GFF_INPUT'] = os.path.join(DIRS['BASE2'], 'gff_output')\n",
    "\n",
    "DIRS['I_FROM_C'] = os.path.join(DIRS['BASE2'], 'bedtools_output', 'intersects_from_cutoffs')\n",
    "DIRS['BED_CUTOFFS'] = os.path.join(DIRS['BASE1'], 'input', 'sorted_bed_files', 'cutoffs')\n",
    "DIRS['6MA_CUTOFFS'] = os.path.join(DIRS['BASE1'], 'input', 'sorted_bed_files', 'cutoffs_6mA')\n",
    "DIRS['5MC_CUTOFFS'] = os.path.join(DIRS['BASE1'], 'input', 'sorted_bed_files', 'cutoffs_5mC')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Making initial dataframes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# make the dataframes for p and h\n",
    "h_contig_df = pd.read_csv(os.path.join(DIRS['FIGURES'], 'expression', 'rnaseq_methyl_df_h.tsv'),sep='\\t' )\n",
    "p_contig_df = pd.read_csv(os.path.join(DIRS['FIGURES'], 'expression', 'rnaseq_methyl_df_p.tsv'),sep='\\t' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>6mA_tss_50</th>\n",
       "      <th>5mC_gene_body_95</th>\n",
       "      <th>6mA_gene_body_95</th>\n",
       "      <th>5mC_upstream_95</th>\n",
       "      <th>6mA_upstream_95</th>\n",
       "      <th>5mC_downstream_95</th>\n",
       "      <th>6mA_downstream_95</th>\n",
       "      <th>5mC_both_95</th>\n",
       "      <th>6mA_both_95</th>\n",
       "      <th>6mA_tss_95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_hcontig_0000_03.1</td>\n",
       "      <td>0.487427</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.377196</td>\n",
       "      <td>0.384531</td>\n",
       "      <td>0.239374</td>\n",
       "      <td>0.147651</td>\n",
       "      <td>0.213</td>\n",
       "      <td>...</td>\n",
       "      <td>0.148</td>\n",
       "      <td>0.105145</td>\n",
       "      <td>0.040268</td>\n",
       "      <td>0.124</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.0835</td>\n",
       "      <td>0.0205</td>\n",
       "      <td>0.040</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_hcontig_0000_03.2</td>\n",
       "      <td>0.577405</td>\n",
       "      <td>144.208849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1016.181477</td>\n",
       "      <td>456.791691</td>\n",
       "      <td>0.446980</td>\n",
       "      <td>0.190050</td>\n",
       "      <td>0.110448</td>\n",
       "      <td>0.176</td>\n",
       "      <td>...</td>\n",
       "      <td>0.092</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_hcontig_0000_03.3</td>\n",
       "      <td>564.783136</td>\n",
       "      <td>52.086774</td>\n",
       "      <td>24.713948</td>\n",
       "      <td>235.779282</td>\n",
       "      <td>239.133947</td>\n",
       "      <td>316.475087</td>\n",
       "      <td>0.172863</td>\n",
       "      <td>0.085502</td>\n",
       "      <td>0.148</td>\n",
       "      <td>...</td>\n",
       "      <td>0.090</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001859</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_hcontig_0000_03.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.167865</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115347</td>\n",
       "      <td>0.179739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.224765</td>\n",
       "      <td>0.177230</td>\n",
       "      <td>0.177</td>\n",
       "      <td>...</td>\n",
       "      <td>0.132</td>\n",
       "      <td>0.001761</td>\n",
       "      <td>0.000587</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.0060</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EVM prediction%2hcontig_0000_003.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.127193</td>\n",
       "      <td>0.245614</td>\n",
       "      <td>0.134</td>\n",
       "      <td>...</td>\n",
       "      <td>0.250</td>\n",
       "      <td>0.000877</td>\n",
       "      <td>0.000877</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>0.002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene_ID          GS          HE        IT0  \\\n",
       "0        gene_model_hcontig_0000_03.1    0.487427    0.000000   0.000000   \n",
       "1        gene_model_hcontig_0000_03.2    0.577405  144.208849   0.000000   \n",
       "2        gene_model_hcontig_0000_03.3  564.783136   52.086774  24.713948   \n",
       "3        gene_model_hcontig_0000_03.4    0.000000    0.167865   0.000000   \n",
       "4  EVM prediction%2hcontig_0000_003.5    0.000000    0.000000   0.000000   \n",
       "\n",
       "           IT6         IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0     0.000000    2.377196    0.384531          0.239374          0.147651   \n",
       "1  1016.181477  456.791691    0.446980          0.190050          0.110448   \n",
       "2   235.779282  239.133947  316.475087          0.172863          0.085502   \n",
       "3     0.115347    0.179739    0.000000          0.224765          0.177230   \n",
       "4     0.000000    0.000000    0.000000          0.127193          0.245614   \n",
       "\n",
       "   5mC_upstream_50     ...      6mA_tss_50  5mC_gene_body_95  \\\n",
       "0            0.213     ...           0.148          0.105145   \n",
       "1            0.176     ...           0.092          0.000000   \n",
       "2            0.148     ...           0.090          0.000000   \n",
       "3            0.177     ...           0.132          0.001761   \n",
       "4            0.134     ...           0.250          0.000877   \n",
       "\n",
       "   6mA_gene_body_95  5mC_upstream_95  6mA_upstream_95  5mC_downstream_95  \\\n",
       "0          0.040268            0.124            0.034              0.043   \n",
       "1          0.000000            0.001            0.000              0.000   \n",
       "2          0.001859            0.000            0.001              0.000   \n",
       "3          0.000587            0.003            0.009              0.000   \n",
       "4          0.000877            0.000            0.003              0.000   \n",
       "\n",
       "   6mA_downstream_95  5mC_both_95  6mA_both_95  6mA_tss_95  \n",
       "0              0.007       0.0835       0.0205       0.040  \n",
       "1              0.002       0.0005       0.0010       0.000  \n",
       "2              0.000       0.0000       0.0005       0.004  \n",
       "3              0.003       0.0015       0.0060       0.002  \n",
       "4              0.003       0.0000       0.0030       0.002  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_contig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>6mA_tss_50</th>\n",
       "      <th>5mC_gene_body_95</th>\n",
       "      <th>6mA_gene_body_95</th>\n",
       "      <th>5mC_upstream_95</th>\n",
       "      <th>6mA_upstream_95</th>\n",
       "      <th>5mC_downstream_95</th>\n",
       "      <th>6mA_downstream_95</th>\n",
       "      <th>5mC_both_95</th>\n",
       "      <th>6mA_both_95</th>\n",
       "      <th>6mA_tss_95</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_pcontig_000.1</td>\n",
       "      <td>9.272092</td>\n",
       "      <td>0.061132</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693499</td>\n",
       "      <td>2.731541</td>\n",
       "      <td>0.232882</td>\n",
       "      <td>0.129151</td>\n",
       "      <td>0.251958</td>\n",
       "      <td>...</td>\n",
       "      <td>0.166</td>\n",
       "      <td>0.070931</td>\n",
       "      <td>0.00779</td>\n",
       "      <td>0.088773</td>\n",
       "      <td>0.016971</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.058886</td>\n",
       "      <td>0.011986</td>\n",
       "      <td>0.01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_pcontig_000.2</td>\n",
       "      <td>0.392567</td>\n",
       "      <td>0.198648</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.925084</td>\n",
       "      <td>0.383660</td>\n",
       "      <td>0.197059</td>\n",
       "      <td>0.120588</td>\n",
       "      <td>0.193000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.128</td>\n",
       "      <td>0.011765</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>0.003000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.001500</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_pcontig_000.3</td>\n",
       "      <td>0.516801</td>\n",
       "      <td>130.454256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>950.078729</td>\n",
       "      <td>419.259148</td>\n",
       "      <td>0.395637</td>\n",
       "      <td>0.198405</td>\n",
       "      <td>0.153539</td>\n",
       "      <td>0.171000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.142</td>\n",
       "      <td>0.000997</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_pcontig_000.4</td>\n",
       "      <td>504.229774</td>\n",
       "      <td>46.756651</td>\n",
       "      <td>21.959973</td>\n",
       "      <td>217.808881</td>\n",
       "      <td>219.104632</td>\n",
       "      <td>279.622402</td>\n",
       "      <td>0.233271</td>\n",
       "      <td>0.091078</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.082</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_model_pcontig_000.5</td>\n",
       "      <td>154.614015</td>\n",
       "      <td>32.043647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>114.040510</td>\n",
       "      <td>450.072586</td>\n",
       "      <td>197.258316</td>\n",
       "      <td>0.232236</td>\n",
       "      <td>0.082611</td>\n",
       "      <td>0.196000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ID          GS          HE        IT0         IT6  \\\n",
       "0  gene_model_pcontig_000.1    9.272092    0.061132   0.000000    0.000000   \n",
       "1  gene_model_pcontig_000.2    0.392567    0.198648   0.000000    0.000000   \n",
       "2  gene_model_pcontig_000.3    0.516801  130.454256   0.000000  950.078729   \n",
       "3  gene_model_pcontig_000.4  504.229774   46.756651  21.959973  217.808881   \n",
       "4  gene_model_pcontig_000.5  154.614015   32.043647   0.000000  114.040510   \n",
       "\n",
       "          IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0    0.693499    2.731541          0.232882          0.129151   \n",
       "1    1.925084    0.383660          0.197059          0.120588   \n",
       "2  419.259148    0.395637          0.198405          0.153539   \n",
       "3  219.104632  279.622402          0.233271          0.091078   \n",
       "4  450.072586  197.258316          0.232236          0.082611   \n",
       "\n",
       "   5mC_upstream_50     ...      6mA_tss_50  5mC_gene_body_95  \\\n",
       "0         0.251958     ...           0.166          0.070931   \n",
       "1         0.193000     ...           0.128          0.011765   \n",
       "2         0.171000     ...           0.142          0.000997   \n",
       "3         0.184000     ...           0.082          0.000000   \n",
       "4         0.196000     ...           0.072          0.000000   \n",
       "\n",
       "   6mA_gene_body_95  5mC_upstream_95  6mA_upstream_95  5mC_downstream_95  \\\n",
       "0           0.00779         0.088773         0.016971              0.029   \n",
       "1           0.00000         0.008000         0.003000              0.000   \n",
       "2           0.00000         0.001000         0.001000              0.000   \n",
       "3           0.00000         0.000000         0.000000              0.000   \n",
       "4           0.00000         0.000000         0.000000              0.000   \n",
       "\n",
       "   6mA_downstream_95  5mC_both_95  6mA_both_95  6mA_tss_95  \n",
       "0              0.007     0.058886     0.011986        0.01  \n",
       "1              0.000     0.004000     0.001500        0.00  \n",
       "2              0.008     0.000500     0.004500        0.00  \n",
       "3              0.000     0.000000     0.000000        0.00  \n",
       "4              0.000     0.000000     0.000000        0.00  \n",
       "\n",
       "[5 rows x 25 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_contig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19348\n",
      "17199\n"
     ]
    }
   ],
   "source": [
    "print(len(p_contig_df))\n",
    "print(len(h_contig_df))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_ID', 'GS', 'HE', 'IT0', 'IT6', 'IT9', 'UG', '5mC_gene_body_50',\n",
       "       '6mA_gene_body_50', '5mC_upstream_50', '6mA_upstream_50',\n",
       "       '5mC_downstream_50', '6mA_downstream_50', '5mC_both_50', '6mA_both_50',\n",
       "       '6mA_tss_50', '5mC_gene_body_95', '6mA_gene_body_95', '5mC_upstream_95',\n",
       "       '6mA_upstream_95', '5mC_downstream_95', '6mA_downstream_95',\n",
       "       '5mC_both_95', '6mA_both_95', '6mA_tss_95'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_contig_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['gene_ID', 'GS', 'HE', 'IT0', 'IT6', 'IT9', 'UG', '5mC_gene_body_50',\n",
       "       '6mA_gene_body_50', '5mC_upstream_50', '6mA_upstream_50',\n",
       "       '5mC_downstream_50', '6mA_downstream_50', '5mC_both_50', '6mA_both_50',\n",
       "       '6mA_tss_50', '5mC_gene_body_95', '6mA_gene_body_95', '5mC_upstream_95',\n",
       "       '6mA_upstream_95', '5mC_downstream_95', '6mA_downstream_95',\n",
       "       '5mC_both_95', '6mA_both_95', '6mA_tss_95'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_contig_df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This adds a protein_ID column at the end of the dataframe for BUSCOs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# This adds a protein_ID column at the end of the dataframe\n",
    "gff_h_anno_TE_df = pd.read_csv('/home/anjuni/analysis/annotations/Pst_104E_v13_h_ctg.repeatgenesLT.gff3', header=None, sep='\\t')\n",
    "gff_h_anno_df = pd.read_csv('/home/anjuni/analysis/annotations/Pst_104E_v13_h_ctg.anno.gff3', header=None, sep='\\t')\n",
    "\n",
    "name_pattern = re.compile(r'Name=([^;]*)')\n",
    "id_pattern = re.compile(r'ID=([^;]*)')\n",
    "\n",
    "Name_id_dict = {}\n",
    "for key in gff_h_anno_df[gff_h_anno_df[2] == 'gene'][8]:\n",
    "    Name_id_dict[re.findall(name_pattern, key)[0]] = re.findall(id_pattern, key)[0]\n",
    "for key in gff_h_anno_TE_df[gff_h_anno_TE_df[2] == 'gene'][8]:\n",
    "    Name_id_dict[re.findall(name_pattern, key)[0]] = re.findall(id_pattern, key)[0]    \n",
    "h_contig_df['Protein_ID'] = h_contig_df['gene_ID'].apply(lambda x: Name_id_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>5mC_gene_body_95</th>\n",
       "      <th>6mA_gene_body_95</th>\n",
       "      <th>5mC_upstream_95</th>\n",
       "      <th>6mA_upstream_95</th>\n",
       "      <th>5mC_downstream_95</th>\n",
       "      <th>6mA_downstream_95</th>\n",
       "      <th>5mC_both_95</th>\n",
       "      <th>6mA_both_95</th>\n",
       "      <th>6mA_tss_95</th>\n",
       "      <th>Protein_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_hcontig_0000_03.1</td>\n",
       "      <td>0.487427</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.377196</td>\n",
       "      <td>0.384531</td>\n",
       "      <td>0.239374</td>\n",
       "      <td>0.147651</td>\n",
       "      <td>0.213</td>\n",
       "      <td>...</td>\n",
       "      <td>0.105145</td>\n",
       "      <td>0.040268</td>\n",
       "      <td>0.124</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.0835</td>\n",
       "      <td>0.0205</td>\n",
       "      <td>0.040</td>\n",
       "      <td>evm.TU.hcontig_000_003.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_hcontig_0000_03.2</td>\n",
       "      <td>0.577405</td>\n",
       "      <td>144.208849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1016.181477</td>\n",
       "      <td>456.791691</td>\n",
       "      <td>0.446980</td>\n",
       "      <td>0.190050</td>\n",
       "      <td>0.110448</td>\n",
       "      <td>0.176</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>0.000</td>\n",
       "      <td>evm.TU.hcontig_000_003.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_hcontig_0000_03.3</td>\n",
       "      <td>564.783136</td>\n",
       "      <td>52.086774</td>\n",
       "      <td>24.713948</td>\n",
       "      <td>235.779282</td>\n",
       "      <td>239.133947</td>\n",
       "      <td>316.475087</td>\n",
       "      <td>0.172863</td>\n",
       "      <td>0.085502</td>\n",
       "      <td>0.148</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001859</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>evm.TU.hcontig_000_003.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_hcontig_0000_03.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.167865</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115347</td>\n",
       "      <td>0.179739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.224765</td>\n",
       "      <td>0.177230</td>\n",
       "      <td>0.177</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001761</td>\n",
       "      <td>0.000587</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.0060</td>\n",
       "      <td>0.002</td>\n",
       "      <td>evm.TU.hcontig_000_003.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EVM prediction%2hcontig_0000_003.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.127193</td>\n",
       "      <td>0.245614</td>\n",
       "      <td>0.134</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000877</td>\n",
       "      <td>0.000877</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>0.002</td>\n",
       "      <td>evm.TU.hcontig_000_003.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene_ID          GS          HE        IT0  \\\n",
       "0        gene_model_hcontig_0000_03.1    0.487427    0.000000   0.000000   \n",
       "1        gene_model_hcontig_0000_03.2    0.577405  144.208849   0.000000   \n",
       "2        gene_model_hcontig_0000_03.3  564.783136   52.086774  24.713948   \n",
       "3        gene_model_hcontig_0000_03.4    0.000000    0.167865   0.000000   \n",
       "4  EVM prediction%2hcontig_0000_003.5    0.000000    0.000000   0.000000   \n",
       "\n",
       "           IT6         IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0     0.000000    2.377196    0.384531          0.239374          0.147651   \n",
       "1  1016.181477  456.791691    0.446980          0.190050          0.110448   \n",
       "2   235.779282  239.133947  316.475087          0.172863          0.085502   \n",
       "3     0.115347    0.179739    0.000000          0.224765          0.177230   \n",
       "4     0.000000    0.000000    0.000000          0.127193          0.245614   \n",
       "\n",
       "   5mC_upstream_50            ...             5mC_gene_body_95  \\\n",
       "0            0.213            ...                     0.105145   \n",
       "1            0.176            ...                     0.000000   \n",
       "2            0.148            ...                     0.000000   \n",
       "3            0.177            ...                     0.001761   \n",
       "4            0.134            ...                     0.000877   \n",
       "\n",
       "   6mA_gene_body_95  5mC_upstream_95  6mA_upstream_95  5mC_downstream_95  \\\n",
       "0          0.040268            0.124            0.034              0.043   \n",
       "1          0.000000            0.001            0.000              0.000   \n",
       "2          0.001859            0.000            0.001              0.000   \n",
       "3          0.000587            0.003            0.009              0.000   \n",
       "4          0.000877            0.000            0.003              0.000   \n",
       "\n",
       "   6mA_downstream_95  5mC_both_95  6mA_both_95  6mA_tss_95  \\\n",
       "0              0.007       0.0835       0.0205       0.040   \n",
       "1              0.002       0.0005       0.0010       0.000   \n",
       "2              0.000       0.0000       0.0005       0.004   \n",
       "3              0.003       0.0015       0.0060       0.002   \n",
       "4              0.003       0.0000       0.0030       0.002   \n",
       "\n",
       "                 Protein_ID  \n",
       "0  evm.TU.hcontig_000_003.1  \n",
       "1  evm.TU.hcontig_000_003.2  \n",
       "2  evm.TU.hcontig_000_003.3  \n",
       "3  evm.TU.hcontig_000_003.4  \n",
       "4  evm.TU.hcontig_000_003.5  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_contig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_p_anno_TE_df = pd.read_csv('/home/anjuni/analysis/annotations/Pst_104E_v13_p_ctg.repeatgenesLT.gff3', header=None, sep='\\t')\n",
    "gff_p_anno_df = pd.read_csv('/home/anjuni/analysis/annotations/Pst_104E_v13_p_ctg.anno.gff3', header=None, sep='\\t')\n",
    "\n",
    "name_pattern = re.compile(r'Name=([^;]*)')\n",
    "id_pattern = re.compile(r'ID=([^;]*)')\n",
    "\n",
    "Name_id_dict = {}\n",
    "for key in gff_p_anno_df[gff_p_anno_df[2] == 'gene'][8]:\n",
    "    Name_id_dict[re.findall(name_pattern, key)[0]] = re.findall(id_pattern, key)[0]\n",
    "for key in gff_p_anno_TE_df[gff_p_anno_TE_df[2] == 'gene'][8]:\n",
    "    Name_id_dict[re.findall(name_pattern, key)[0]] = re.findall(id_pattern, key)[0]\n",
    "p_contig_df['Protein_ID'] = p_contig_df['gene_ID'].apply(lambda x: Name_id_dict[x])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>5mC_gene_body_95</th>\n",
       "      <th>6mA_gene_body_95</th>\n",
       "      <th>5mC_upstream_95</th>\n",
       "      <th>6mA_upstream_95</th>\n",
       "      <th>5mC_downstream_95</th>\n",
       "      <th>6mA_downstream_95</th>\n",
       "      <th>5mC_both_95</th>\n",
       "      <th>6mA_both_95</th>\n",
       "      <th>6mA_tss_95</th>\n",
       "      <th>Protein_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_pcontig_000.1</td>\n",
       "      <td>9.272092</td>\n",
       "      <td>0.061132</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693499</td>\n",
       "      <td>2.731541</td>\n",
       "      <td>0.232882</td>\n",
       "      <td>0.129151</td>\n",
       "      <td>0.251958</td>\n",
       "      <td>...</td>\n",
       "      <td>0.070931</td>\n",
       "      <td>0.00779</td>\n",
       "      <td>0.088773</td>\n",
       "      <td>0.016971</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.058886</td>\n",
       "      <td>0.011986</td>\n",
       "      <td>0.01</td>\n",
       "      <td>evm.TU.pcontig_000.1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_pcontig_000.2</td>\n",
       "      <td>0.392567</td>\n",
       "      <td>0.198648</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.925084</td>\n",
       "      <td>0.383660</td>\n",
       "      <td>0.197059</td>\n",
       "      <td>0.120588</td>\n",
       "      <td>0.193000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.011765</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>0.003000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.001500</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_pcontig_000.3</td>\n",
       "      <td>0.516801</td>\n",
       "      <td>130.454256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>950.078729</td>\n",
       "      <td>419.259148</td>\n",
       "      <td>0.395637</td>\n",
       "      <td>0.198405</td>\n",
       "      <td>0.153539</td>\n",
       "      <td>0.171000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000997</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_pcontig_000.4</td>\n",
       "      <td>504.229774</td>\n",
       "      <td>46.756651</td>\n",
       "      <td>21.959973</td>\n",
       "      <td>217.808881</td>\n",
       "      <td>219.104632</td>\n",
       "      <td>279.622402</td>\n",
       "      <td>0.233271</td>\n",
       "      <td>0.091078</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_model_pcontig_000.5</td>\n",
       "      <td>154.614015</td>\n",
       "      <td>32.043647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>114.040510</td>\n",
       "      <td>450.072586</td>\n",
       "      <td>197.258316</td>\n",
       "      <td>0.232236</td>\n",
       "      <td>0.082611</td>\n",
       "      <td>0.196000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ID          GS          HE        IT0         IT6  \\\n",
       "0  gene_model_pcontig_000.1    9.272092    0.061132   0.000000    0.000000   \n",
       "1  gene_model_pcontig_000.2    0.392567    0.198648   0.000000    0.000000   \n",
       "2  gene_model_pcontig_000.3    0.516801  130.454256   0.000000  950.078729   \n",
       "3  gene_model_pcontig_000.4  504.229774   46.756651  21.959973  217.808881   \n",
       "4  gene_model_pcontig_000.5  154.614015   32.043647   0.000000  114.040510   \n",
       "\n",
       "          IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0    0.693499    2.731541          0.232882          0.129151   \n",
       "1    1.925084    0.383660          0.197059          0.120588   \n",
       "2  419.259148    0.395637          0.198405          0.153539   \n",
       "3  219.104632  279.622402          0.233271          0.091078   \n",
       "4  450.072586  197.258316          0.232236          0.082611   \n",
       "\n",
       "   5mC_upstream_50          ...           5mC_gene_body_95  6mA_gene_body_95  \\\n",
       "0         0.251958          ...                   0.070931           0.00779   \n",
       "1         0.193000          ...                   0.011765           0.00000   \n",
       "2         0.171000          ...                   0.000997           0.00000   \n",
       "3         0.184000          ...                   0.000000           0.00000   \n",
       "4         0.196000          ...                   0.000000           0.00000   \n",
       "\n",
       "   5mC_upstream_95  6mA_upstream_95  5mC_downstream_95  6mA_downstream_95  \\\n",
       "0         0.088773         0.016971              0.029              0.007   \n",
       "1         0.008000         0.003000              0.000              0.000   \n",
       "2         0.001000         0.001000              0.000              0.008   \n",
       "3         0.000000         0.000000              0.000              0.000   \n",
       "4         0.000000         0.000000              0.000              0.000   \n",
       "\n",
       "   5mC_both_95  6mA_both_95  6mA_tss_95            Protein_ID  \n",
       "0     0.058886     0.011986        0.01  evm.TU.pcontig_000.1  \n",
       "1     0.004000     0.001500        0.00  evm.TU.pcontig_000.2  \n",
       "2     0.000500     0.004500        0.00  evm.TU.pcontig_000.3  \n",
       "3     0.000000     0.000000        0.00  evm.TU.pcontig_000.4  \n",
       "4     0.000000     0.000000        0.00  evm.TU.pcontig_000.5  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_contig_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This adds a column for gene type at the end of the dataframe for gene type\n",
    "- BUSCOs, effectors, TEs, all genes, non-TE genes, all genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a list of genes in each gene type category and add the column\n",
    "\n",
    "#add file paths\n",
    "GENE_BED_FOLDER = '/home/anjuni/beds/'\n",
    "busco_gene_bed_h = 'Pst_104E_v12_h_busco.gene.bed'\n",
    "effector_gene_bed_h = 'Pst_104E_v12_h_effector.gene.bed'\n",
    "TE_list_h = []\n",
    "for key in gff_h_anno_TE_df[gff_h_anno_TE_df[2] == 'gene'][8]:\n",
    "    TE_list_h.append(re.findall(id_pattern, key)[0])\n",
    "\n",
    "# make a list\n",
    "BUSCO_list_h = pd.read_csv(os.path.join(GENE_BED_FOLDER, busco_gene_bed_h), sep='\\t', header=None)[3].tolist()\n",
    "effector_list_h = pd.read_csv(os.path.join(GENE_BED_FOLDER, effector_gene_bed_h), sep='\\t', header=None)[3].tolist()\n",
    "\n",
    "# add values to column\n",
    "h_contig_df['Gene_type'] = 'All genes'\n",
    "h_contig_df.loc[h_contig_df.Protein_ID.isin(effector_list_h), 'Gene_type'] = 'Effector'\n",
    "h_contig_df.loc[h_contig_df.Protein_ID.isin(BUSCO_list_h), 'Gene_type'] = 'BUSCO'\n",
    "h_contig_df.loc[h_contig_df.Protein_ID.isin(TE_list_h), 'Gene_type'] = 'TE gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>6mA_gene_body_95</th>\n",
       "      <th>5mC_upstream_95</th>\n",
       "      <th>6mA_upstream_95</th>\n",
       "      <th>5mC_downstream_95</th>\n",
       "      <th>6mA_downstream_95</th>\n",
       "      <th>5mC_both_95</th>\n",
       "      <th>6mA_both_95</th>\n",
       "      <th>6mA_tss_95</th>\n",
       "      <th>Protein_ID</th>\n",
       "      <th>Gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_hcontig_0000_03.1</td>\n",
       "      <td>0.487427</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.377196</td>\n",
       "      <td>0.384531</td>\n",
       "      <td>0.239374</td>\n",
       "      <td>0.147651</td>\n",
       "      <td>0.213</td>\n",
       "      <td>...</td>\n",
       "      <td>0.040268</td>\n",
       "      <td>0.124</td>\n",
       "      <td>0.034</td>\n",
       "      <td>0.043</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.0835</td>\n",
       "      <td>0.0205</td>\n",
       "      <td>0.040</td>\n",
       "      <td>evm.TU.hcontig_000_003.1</td>\n",
       "      <td>Effector</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_hcontig_0000_03.2</td>\n",
       "      <td>0.577405</td>\n",
       "      <td>144.208849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1016.181477</td>\n",
       "      <td>456.791691</td>\n",
       "      <td>0.446980</td>\n",
       "      <td>0.190050</td>\n",
       "      <td>0.110448</td>\n",
       "      <td>0.176</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.002</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>0.000</td>\n",
       "      <td>evm.TU.hcontig_000_003.2</td>\n",
       "      <td>Effector</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_hcontig_0000_03.3</td>\n",
       "      <td>564.783136</td>\n",
       "      <td>52.086774</td>\n",
       "      <td>24.713948</td>\n",
       "      <td>235.779282</td>\n",
       "      <td>239.133947</td>\n",
       "      <td>316.475087</td>\n",
       "      <td>0.172863</td>\n",
       "      <td>0.085502</td>\n",
       "      <td>0.148</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001859</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.001</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>evm.TU.hcontig_000_003.3</td>\n",
       "      <td>All genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_hcontig_0000_03.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.167865</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115347</td>\n",
       "      <td>0.179739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.224765</td>\n",
       "      <td>0.177230</td>\n",
       "      <td>0.177</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000587</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.0060</td>\n",
       "      <td>0.002</td>\n",
       "      <td>evm.TU.hcontig_000_003.4</td>\n",
       "      <td>All genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EVM prediction%2hcontig_0000_003.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.127193</td>\n",
       "      <td>0.245614</td>\n",
       "      <td>0.134</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000877</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>0.002</td>\n",
       "      <td>evm.TU.hcontig_000_003.5</td>\n",
       "      <td>TE gene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene_ID          GS          HE        IT0  \\\n",
       "0        gene_model_hcontig_0000_03.1    0.487427    0.000000   0.000000   \n",
       "1        gene_model_hcontig_0000_03.2    0.577405  144.208849   0.000000   \n",
       "2        gene_model_hcontig_0000_03.3  564.783136   52.086774  24.713948   \n",
       "3        gene_model_hcontig_0000_03.4    0.000000    0.167865   0.000000   \n",
       "4  EVM prediction%2hcontig_0000_003.5    0.000000    0.000000   0.000000   \n",
       "\n",
       "           IT6         IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0     0.000000    2.377196    0.384531          0.239374          0.147651   \n",
       "1  1016.181477  456.791691    0.446980          0.190050          0.110448   \n",
       "2   235.779282  239.133947  316.475087          0.172863          0.085502   \n",
       "3     0.115347    0.179739    0.000000          0.224765          0.177230   \n",
       "4     0.000000    0.000000    0.000000          0.127193          0.245614   \n",
       "\n",
       "   5mC_upstream_50    ...      6mA_gene_body_95  5mC_upstream_95  \\\n",
       "0            0.213    ...              0.040268            0.124   \n",
       "1            0.176    ...              0.000000            0.001   \n",
       "2            0.148    ...              0.001859            0.000   \n",
       "3            0.177    ...              0.000587            0.003   \n",
       "4            0.134    ...              0.000877            0.000   \n",
       "\n",
       "   6mA_upstream_95  5mC_downstream_95  6mA_downstream_95  5mC_both_95  \\\n",
       "0            0.034              0.043              0.007       0.0835   \n",
       "1            0.000              0.000              0.002       0.0005   \n",
       "2            0.001              0.000              0.000       0.0000   \n",
       "3            0.009              0.000              0.003       0.0015   \n",
       "4            0.003              0.000              0.003       0.0000   \n",
       "\n",
       "   6mA_both_95  6mA_tss_95                Protein_ID  Gene_type  \n",
       "0       0.0205       0.040  evm.TU.hcontig_000_003.1   Effector  \n",
       "1       0.0010       0.000  evm.TU.hcontig_000_003.2   Effector  \n",
       "2       0.0005       0.004  evm.TU.hcontig_000_003.3  All genes  \n",
       "3       0.0060       0.002  evm.TU.hcontig_000_003.4  All genes  \n",
       "4       0.0030       0.002  evm.TU.hcontig_000_003.5    TE gene  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_contig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get a list of genes in each gene type category and add the column\n",
    "\n",
    "#add file paths\n",
    "\n",
    "busco_gene_bed_p = 'Pst_104E_v12_p_busco.gene.bed'\n",
    "effector_gene_bed_p = 'Pst_104E_v12_p_effector.gene.bed'\n",
    "TE_list_p = []\n",
    "for key in gff_p_anno_TE_df[gff_p_anno_TE_df[2] == 'gene'][8]:\n",
    "    TE_list_p.append(re.findall(id_pattern, key)[0])\n",
    "#s.listdir(GENE_BED_FOLDER)\n",
    "\n",
    "# make a list\n",
    "BUSCO_list_p = pd.read_csv(os.path.join(GENE_BED_FOLDER, busco_gene_bed_p), sep='\\t', header=None)[3].tolist()\n",
    "effector_list_p = pd.read_csv(os.path.join(GENE_BED_FOLDER, effector_gene_bed_p), sep='\\t', header=None)[3].tolist()\n",
    "\n",
    "# add values to column\n",
    "p_contig_df['Gene_type'] = 'All genes'\n",
    "p_contig_df.loc[p_contig_df.Protein_ID.isin(effector_list_p), 'Gene_type'] = 'Effector'\n",
    "p_contig_df.loc[p_contig_df.Protein_ID.isin(BUSCO_list_p), 'Gene_type'] = 'BUSCO'\n",
    "p_contig_df.loc[p_contig_df.Protein_ID.isin(TE_list_p), 'Gene_type'] = 'TE gene'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>6mA_gene_body_95</th>\n",
       "      <th>5mC_upstream_95</th>\n",
       "      <th>6mA_upstream_95</th>\n",
       "      <th>5mC_downstream_95</th>\n",
       "      <th>6mA_downstream_95</th>\n",
       "      <th>5mC_both_95</th>\n",
       "      <th>6mA_both_95</th>\n",
       "      <th>6mA_tss_95</th>\n",
       "      <th>Protein_ID</th>\n",
       "      <th>Gene_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_pcontig_000.1</td>\n",
       "      <td>9.272092</td>\n",
       "      <td>0.061132</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693499</td>\n",
       "      <td>2.731541</td>\n",
       "      <td>0.232882</td>\n",
       "      <td>0.129151</td>\n",
       "      <td>0.251958</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00779</td>\n",
       "      <td>0.088773</td>\n",
       "      <td>0.016971</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.058886</td>\n",
       "      <td>0.011986</td>\n",
       "      <td>0.01</td>\n",
       "      <td>evm.TU.pcontig_000.1</td>\n",
       "      <td>All genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_pcontig_000.2</td>\n",
       "      <td>0.392567</td>\n",
       "      <td>0.198648</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.925084</td>\n",
       "      <td>0.383660</td>\n",
       "      <td>0.197059</td>\n",
       "      <td>0.120588</td>\n",
       "      <td>0.193000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.008000</td>\n",
       "      <td>0.003000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.001500</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.2</td>\n",
       "      <td>All genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_pcontig_000.3</td>\n",
       "      <td>0.516801</td>\n",
       "      <td>130.454256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>950.078729</td>\n",
       "      <td>419.259148</td>\n",
       "      <td>0.395637</td>\n",
       "      <td>0.198405</td>\n",
       "      <td>0.153539</td>\n",
       "      <td>0.171000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.008</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.3</td>\n",
       "      <td>Effector</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_pcontig_000.4</td>\n",
       "      <td>504.229774</td>\n",
       "      <td>46.756651</td>\n",
       "      <td>21.959973</td>\n",
       "      <td>217.808881</td>\n",
       "      <td>219.104632</td>\n",
       "      <td>279.622402</td>\n",
       "      <td>0.233271</td>\n",
       "      <td>0.091078</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.4</td>\n",
       "      <td>All genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_model_pcontig_000.5</td>\n",
       "      <td>154.614015</td>\n",
       "      <td>32.043647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>114.040510</td>\n",
       "      <td>450.072586</td>\n",
       "      <td>197.258316</td>\n",
       "      <td>0.232236</td>\n",
       "      <td>0.082611</td>\n",
       "      <td>0.196000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.5</td>\n",
       "      <td>All genes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ID          GS          HE        IT0         IT6  \\\n",
       "0  gene_model_pcontig_000.1    9.272092    0.061132   0.000000    0.000000   \n",
       "1  gene_model_pcontig_000.2    0.392567    0.198648   0.000000    0.000000   \n",
       "2  gene_model_pcontig_000.3    0.516801  130.454256   0.000000  950.078729   \n",
       "3  gene_model_pcontig_000.4  504.229774   46.756651  21.959973  217.808881   \n",
       "4  gene_model_pcontig_000.5  154.614015   32.043647   0.000000  114.040510   \n",
       "\n",
       "          IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0    0.693499    2.731541          0.232882          0.129151   \n",
       "1    1.925084    0.383660          0.197059          0.120588   \n",
       "2  419.259148    0.395637          0.198405          0.153539   \n",
       "3  219.104632  279.622402          0.233271          0.091078   \n",
       "4  450.072586  197.258316          0.232236          0.082611   \n",
       "\n",
       "   5mC_upstream_50    ...      6mA_gene_body_95  5mC_upstream_95  \\\n",
       "0         0.251958    ...               0.00779         0.088773   \n",
       "1         0.193000    ...               0.00000         0.008000   \n",
       "2         0.171000    ...               0.00000         0.001000   \n",
       "3         0.184000    ...               0.00000         0.000000   \n",
       "4         0.196000    ...               0.00000         0.000000   \n",
       "\n",
       "   6mA_upstream_95  5mC_downstream_95  6mA_downstream_95  5mC_both_95  \\\n",
       "0         0.016971              0.029              0.007     0.058886   \n",
       "1         0.003000              0.000              0.000     0.004000   \n",
       "2         0.001000              0.000              0.008     0.000500   \n",
       "3         0.000000              0.000              0.000     0.000000   \n",
       "4         0.000000              0.000              0.000     0.000000   \n",
       "\n",
       "   6mA_both_95  6mA_tss_95            Protein_ID  Gene_type  \n",
       "0     0.011986        0.01  evm.TU.pcontig_000.1  All genes  \n",
       "1     0.001500        0.00  evm.TU.pcontig_000.2  All genes  \n",
       "2     0.004500        0.00  evm.TU.pcontig_000.3   Effector  \n",
       "3     0.000000        0.00  evm.TU.pcontig_000.4  All genes  \n",
       "4     0.000000        0.00  evm.TU.pcontig_000.5  All genes  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_contig_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## This adds a columns at the end of the dataframe for the top 500 genes for each methylation type\n",
    "\n",
    "- top 500 5mc gene body\n",
    "- top 500 5mc upstream\n",
    "- top 500 5mc downstream\n",
    "- top 500 5mc both\n",
    "- top 500 6ma tss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "###This is the index of all highly methylated genes\n",
    "#get a true false column for plotting and analysis\n",
    "h_contig_df[\"Top_500_5mC_gene_body_50\"] = \"Lowly methylated\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='5mC_gene_body_50', ascending=False).loc[:500].index,\"Top_500_5mC_gene_body_50\"] = \"Highly methylated\"\n",
    "\n",
    "h_contig_df[\"Top_500_5mC_upstream_50\"] = \"Lowly methylated\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='5mC_upstream_50', ascending=False).loc[:500].index,\"Top_500_5mC_upstream_50\"] = \"Highly methylated\"\n",
    "\n",
    "h_contig_df[\"Top_500_5mC_downstream_50\"] = \"Lowly methylated\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='5mC_downstream_50', ascending=False).loc[:500].index,\"Top_500_5mC_downstream_50\"] = \"Highly methylated\"\n",
    "\n",
    "h_contig_df[\"Top_500_5mC_both_50\"] = \"Lowly methylated\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='5mC_both_50', ascending=False).loc[:500].index,\"Top_500_5mC_both_50\"] = \"Highly methylated\"\n",
    "\n",
    "h_contig_df[\"Top_500_6mA_tss_50\"] = \"Lowly methylated\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='6mA_tss_50', ascending=False).loc[:500].index,\"Top_500_6mA_tss_50\"] = \"Highly methylated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "###This is the index of all highly methylated genes\n",
    "#get a true false column for plotting and analysis\n",
    "h_contig_df[\"Top_500_5mC_gene_body_50\"] = \"Low\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='5mC_gene_body_50', ascending=False).loc[:500].index,\"Top_500_5mC_gene_body_50\"] = \"High\"\n",
    "\n",
    "h_contig_df[\"Top_500_5mC_upstream_50\"] = \"Low\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='5mC_upstream_50', ascending=False).loc[:500].index,\"Top_500_5mC_upstream_50\"] = \"High\"\n",
    "\n",
    "h_contig_df[\"Top_500_5mC_downstream_50\"] = \"Low\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='5mC_downstream_50', ascending=False).loc[:500].index,\"Top_500_5mC_downstream_50\"] = \"High\"\n",
    "\n",
    "h_contig_df[\"Top_500_5mC_both_50\"] = \"Low\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='5mC_both_50', ascending=False).loc[:500].index,\"Top_500_5mC_both_50\"] = \"High\"\n",
    "\n",
    "h_contig_df[\"Top_500_6mA_tss_50\"] = \"Low\"\n",
    "h_contig_df.loc[h_contig_df.sort_values(by='6mA_tss_50', ascending=False).loc[:500].index,\"Top_500_6mA_tss_50\"] = \"High\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>5mC_both_95</th>\n",
       "      <th>6mA_both_95</th>\n",
       "      <th>6mA_tss_95</th>\n",
       "      <th>Protein_ID</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Top_500_5mC_gene_body_50</th>\n",
       "      <th>Top_500_5mC_upstream_50</th>\n",
       "      <th>Top_500_5mC_downstream_50</th>\n",
       "      <th>Top_500_5mC_both_50</th>\n",
       "      <th>Top_500_6mA_tss_50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_hcontig_0000_03.1</td>\n",
       "      <td>0.487427</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.377196</td>\n",
       "      <td>0.384531</td>\n",
       "      <td>0.239374</td>\n",
       "      <td>0.147651</td>\n",
       "      <td>0.213</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0835</td>\n",
       "      <td>0.0205</td>\n",
       "      <td>0.040</td>\n",
       "      <td>evm.TU.hcontig_000_003.1</td>\n",
       "      <td>Effector</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_hcontig_0000_03.2</td>\n",
       "      <td>0.577405</td>\n",
       "      <td>144.208849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1016.181477</td>\n",
       "      <td>456.791691</td>\n",
       "      <td>0.446980</td>\n",
       "      <td>0.190050</td>\n",
       "      <td>0.110448</td>\n",
       "      <td>0.176</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>0.000</td>\n",
       "      <td>evm.TU.hcontig_000_003.2</td>\n",
       "      <td>Effector</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_hcontig_0000_03.3</td>\n",
       "      <td>564.783136</td>\n",
       "      <td>52.086774</td>\n",
       "      <td>24.713948</td>\n",
       "      <td>235.779282</td>\n",
       "      <td>239.133947</td>\n",
       "      <td>316.475087</td>\n",
       "      <td>0.172863</td>\n",
       "      <td>0.085502</td>\n",
       "      <td>0.148</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0005</td>\n",
       "      <td>0.004</td>\n",
       "      <td>evm.TU.hcontig_000_003.3</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_hcontig_0000_03.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.167865</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115347</td>\n",
       "      <td>0.179739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.224765</td>\n",
       "      <td>0.177230</td>\n",
       "      <td>0.177</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0.0060</td>\n",
       "      <td>0.002</td>\n",
       "      <td>evm.TU.hcontig_000_003.4</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EVM prediction%2hcontig_0000_003.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.127193</td>\n",
       "      <td>0.245614</td>\n",
       "      <td>0.134</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0030</td>\n",
       "      <td>0.002</td>\n",
       "      <td>evm.TU.hcontig_000_003.5</td>\n",
       "      <td>TE gene</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene_ID          GS          HE        IT0  \\\n",
       "0        gene_model_hcontig_0000_03.1    0.487427    0.000000   0.000000   \n",
       "1        gene_model_hcontig_0000_03.2    0.577405  144.208849   0.000000   \n",
       "2        gene_model_hcontig_0000_03.3  564.783136   52.086774  24.713948   \n",
       "3        gene_model_hcontig_0000_03.4    0.000000    0.167865   0.000000   \n",
       "4  EVM prediction%2hcontig_0000_003.5    0.000000    0.000000   0.000000   \n",
       "\n",
       "           IT6         IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0     0.000000    2.377196    0.384531          0.239374          0.147651   \n",
       "1  1016.181477  456.791691    0.446980          0.190050          0.110448   \n",
       "2   235.779282  239.133947  316.475087          0.172863          0.085502   \n",
       "3     0.115347    0.179739    0.000000          0.224765          0.177230   \n",
       "4     0.000000    0.000000    0.000000          0.127193          0.245614   \n",
       "\n",
       "   5mC_upstream_50         ...          5mC_both_95  6mA_both_95  6mA_tss_95  \\\n",
       "0            0.213         ...               0.0835       0.0205       0.040   \n",
       "1            0.176         ...               0.0005       0.0010       0.000   \n",
       "2            0.148         ...               0.0000       0.0005       0.004   \n",
       "3            0.177         ...               0.0015       0.0060       0.002   \n",
       "4            0.134         ...               0.0000       0.0030       0.002   \n",
       "\n",
       "                 Protein_ID  Gene_type  Top_500_5mC_gene_body_50  \\\n",
       "0  evm.TU.hcontig_000_003.1   Effector                      High   \n",
       "1  evm.TU.hcontig_000_003.2   Effector                      High   \n",
       "2  evm.TU.hcontig_000_003.3  All genes                      High   \n",
       "3  evm.TU.hcontig_000_003.4  All genes                      High   \n",
       "4  evm.TU.hcontig_000_003.5    TE gene                      High   \n",
       "\n",
       "   Top_500_5mC_upstream_50  Top_500_5mC_downstream_50  Top_500_5mC_both_50  \\\n",
       "0                     High                       High                 High   \n",
       "1                     High                       High                 High   \n",
       "2                     High                       High                 High   \n",
       "3                     High                       High                 High   \n",
       "4                     High                       High                 High   \n",
       "\n",
       "   Top_500_6mA_tss_50  \n",
       "0                High  \n",
       "1                High  \n",
       "2                High  \n",
       "3                High  \n",
       "4                High  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_contig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "###This is the index of all highly methylated genes\n",
    "#get a true false column for plotting and analysis\n",
    "p_contig_df[\"Top_500_5mC_gene_body_50\"] = \"Lowly methylated\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='5mC_gene_body_50', ascending=False).loc[:500].index,\"Top_500_5mC_gene_body_50\"] = \"Highly methylated\"\n",
    "\n",
    "p_contig_df[\"Top_500_5mC_upstream_50\"] = \"Lowly methylated\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='5mC_upstream_50', ascending=False).loc[:500].index,\"Top_500_5mC_upstream_50\"] = \"Highly methylated\"\n",
    "\n",
    "p_contig_df[\"Top_500_5mC_downstream_50\"] = \"Lowly methylated\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='5mC_downstream_50', ascending=False).loc[:500].index,\"Top_500_5mC_downstream_50\"] = \"Highly methylated\"\n",
    "\n",
    "p_contig_df[\"Top_500_5mC_both_50\"] = \"Lowly methylated\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='5mC_both_50', ascending=False).loc[:500].index,\"Top_500_5mC_both_50\"] = \"Highly methylated\"\n",
    "\n",
    "p_contig_df[\"Top_500_6mA_tss_50\"] = \"Lowly methylated\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='6mA_tss_50', ascending=False).loc[:500].index,\"Top_500_6mA_tss_50\"] = \"Highly methylated\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "###This is the index of all highly methylated genes\n",
    "#get a true false column for plotting and analysis\n",
    "p_contig_df[\"Top_500_5mC_gene_body_50\"] = \"Low\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='5mC_gene_body_50', ascending=False).loc[:500].index,\"Top_500_5mC_gene_body_50\"] = \"High\"\n",
    "\n",
    "p_contig_df[\"Top_500_5mC_upstream_50\"] = \"Low\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='5mC_upstream_50', ascending=False).loc[:500].index,\"Top_500_5mC_upstream_50\"] = \"High\"\n",
    "\n",
    "p_contig_df[\"Top_500_5mC_downstream_50\"] = \"Low\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='5mC_downstream_50', ascending=False).loc[:500].index,\"Top_500_5mC_downstream_50\"] = \"High\"\n",
    "\n",
    "p_contig_df[\"Top_500_5mC_both_50\"] = \"Low\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='5mC_both_50', ascending=False).loc[:500].index,\"Top_500_5mC_both_50\"] = \"High\"\n",
    "\n",
    "p_contig_df[\"Top_500_6mA_tss_50\"] = \"Low\"\n",
    "p_contig_df.loc[p_contig_df.sort_values(by='6mA_tss_50', ascending=False).loc[:500].index,\"Top_500_6mA_tss_50\"] = \"High\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>5mC_both_95</th>\n",
       "      <th>6mA_both_95</th>\n",
       "      <th>6mA_tss_95</th>\n",
       "      <th>Protein_ID</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Top_500_5mC_gene_body_50</th>\n",
       "      <th>Top_500_5mC_upstream_50</th>\n",
       "      <th>Top_500_5mC_downstream_50</th>\n",
       "      <th>Top_500_5mC_both_50</th>\n",
       "      <th>Top_500_6mA_tss_50</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_pcontig_000.1</td>\n",
       "      <td>9.272092</td>\n",
       "      <td>0.061132</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693499</td>\n",
       "      <td>2.731541</td>\n",
       "      <td>0.232882</td>\n",
       "      <td>0.129151</td>\n",
       "      <td>0.251958</td>\n",
       "      <td>...</td>\n",
       "      <td>0.058886</td>\n",
       "      <td>0.011986</td>\n",
       "      <td>0.01</td>\n",
       "      <td>evm.TU.pcontig_000.1</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_pcontig_000.2</td>\n",
       "      <td>0.392567</td>\n",
       "      <td>0.198648</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.925084</td>\n",
       "      <td>0.383660</td>\n",
       "      <td>0.197059</td>\n",
       "      <td>0.120588</td>\n",
       "      <td>0.193000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.004000</td>\n",
       "      <td>0.001500</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.2</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_pcontig_000.3</td>\n",
       "      <td>0.516801</td>\n",
       "      <td>130.454256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>950.078729</td>\n",
       "      <td>419.259148</td>\n",
       "      <td>0.395637</td>\n",
       "      <td>0.198405</td>\n",
       "      <td>0.153539</td>\n",
       "      <td>0.171000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000500</td>\n",
       "      <td>0.004500</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.3</td>\n",
       "      <td>Effector</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_pcontig_000.4</td>\n",
       "      <td>504.229774</td>\n",
       "      <td>46.756651</td>\n",
       "      <td>21.959973</td>\n",
       "      <td>217.808881</td>\n",
       "      <td>219.104632</td>\n",
       "      <td>279.622402</td>\n",
       "      <td>0.233271</td>\n",
       "      <td>0.091078</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.4</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_model_pcontig_000.5</td>\n",
       "      <td>154.614015</td>\n",
       "      <td>32.043647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>114.040510</td>\n",
       "      <td>450.072586</td>\n",
       "      <td>197.258316</td>\n",
       "      <td>0.232236</td>\n",
       "      <td>0.082611</td>\n",
       "      <td>0.196000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>evm.TU.pcontig_000.5</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 32 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ID          GS          HE        IT0         IT6  \\\n",
       "0  gene_model_pcontig_000.1    9.272092    0.061132   0.000000    0.000000   \n",
       "1  gene_model_pcontig_000.2    0.392567    0.198648   0.000000    0.000000   \n",
       "2  gene_model_pcontig_000.3    0.516801  130.454256   0.000000  950.078729   \n",
       "3  gene_model_pcontig_000.4  504.229774   46.756651  21.959973  217.808881   \n",
       "4  gene_model_pcontig_000.5  154.614015   32.043647   0.000000  114.040510   \n",
       "\n",
       "          IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0    0.693499    2.731541          0.232882          0.129151   \n",
       "1    1.925084    0.383660          0.197059          0.120588   \n",
       "2  419.259148    0.395637          0.198405          0.153539   \n",
       "3  219.104632  279.622402          0.233271          0.091078   \n",
       "4  450.072586  197.258316          0.232236          0.082611   \n",
       "\n",
       "   5mC_upstream_50         ...          5mC_both_95  6mA_both_95  6mA_tss_95  \\\n",
       "0         0.251958         ...             0.058886     0.011986        0.01   \n",
       "1         0.193000         ...             0.004000     0.001500        0.00   \n",
       "2         0.171000         ...             0.000500     0.004500        0.00   \n",
       "3         0.184000         ...             0.000000     0.000000        0.00   \n",
       "4         0.196000         ...             0.000000     0.000000        0.00   \n",
       "\n",
       "             Protein_ID  Gene_type  Top_500_5mC_gene_body_50  \\\n",
       "0  evm.TU.pcontig_000.1  All genes                      High   \n",
       "1  evm.TU.pcontig_000.2  All genes                      High   \n",
       "2  evm.TU.pcontig_000.3   Effector                      High   \n",
       "3  evm.TU.pcontig_000.4  All genes                      High   \n",
       "4  evm.TU.pcontig_000.5  All genes                      High   \n",
       "\n",
       "   Top_500_5mC_upstream_50  Top_500_5mC_downstream_50  Top_500_5mC_both_50  \\\n",
       "0                     High                       High                 High   \n",
       "1                     High                       High                 High   \n",
       "2                     High                       High                 High   \n",
       "3                     High                       High                 High   \n",
       "4                     High                       High                 High   \n",
       "\n",
       "   Top_500_6mA_tss_50  \n",
       "0                High  \n",
       "1                High  \n",
       "2                High  \n",
       "3                High  \n",
       "4                High  \n",
       "\n",
       "[5 rows x 32 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_contig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#add 0.1 for log transformation\n",
    "h_contig_df['UG_01'] = h_contig_df['UG'] + 0.1\n",
    "h_contig_df['HE_01'] = h_contig_df['HE'] + 0.1\n",
    "h_contig_df['IT6_01'] = h_contig_df['IT6'] + 0.1\n",
    "h_contig_df['IT9_01'] = h_contig_df['IT9'] + 0.1\n",
    "\n",
    "p_contig_df['UG_01'] = h_contig_df['UG'] + 0.1\n",
    "p_contig_df['HE_01'] = h_contig_df['HE'] + 0.1\n",
    "p_contig_df['IT6_01'] = h_contig_df['IT6'] + 0.1\n",
    "p_contig_df['IT9_01'] = h_contig_df['IT9'] + 0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Top_500_5mC_gene_body_50</th>\n",
       "      <th>Top_500_5mC_upstream_50</th>\n",
       "      <th>Top_500_5mC_downstream_50</th>\n",
       "      <th>Top_500_5mC_both_50</th>\n",
       "      <th>Top_500_6mA_tss_50</th>\n",
       "      <th>UG_01</th>\n",
       "      <th>HE_01</th>\n",
       "      <th>IT6_01</th>\n",
       "      <th>IT9_01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_hcontig_0000_03.1</td>\n",
       "      <td>0.487427</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.377196</td>\n",
       "      <td>0.384531</td>\n",
       "      <td>0.239374</td>\n",
       "      <td>0.147651</td>\n",
       "      <td>0.213</td>\n",
       "      <td>...</td>\n",
       "      <td>Effector</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.484531</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>2.477196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_hcontig_0000_03.2</td>\n",
       "      <td>0.577405</td>\n",
       "      <td>144.208849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1016.181477</td>\n",
       "      <td>456.791691</td>\n",
       "      <td>0.446980</td>\n",
       "      <td>0.190050</td>\n",
       "      <td>0.110448</td>\n",
       "      <td>0.176</td>\n",
       "      <td>...</td>\n",
       "      <td>Effector</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.546980</td>\n",
       "      <td>144.308849</td>\n",
       "      <td>1016.281477</td>\n",
       "      <td>456.891691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_hcontig_0000_03.3</td>\n",
       "      <td>564.783136</td>\n",
       "      <td>52.086774</td>\n",
       "      <td>24.713948</td>\n",
       "      <td>235.779282</td>\n",
       "      <td>239.133947</td>\n",
       "      <td>316.475087</td>\n",
       "      <td>0.172863</td>\n",
       "      <td>0.085502</td>\n",
       "      <td>0.148</td>\n",
       "      <td>...</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>316.575087</td>\n",
       "      <td>52.186774</td>\n",
       "      <td>235.879282</td>\n",
       "      <td>239.233947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_hcontig_0000_03.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.167865</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115347</td>\n",
       "      <td>0.179739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.224765</td>\n",
       "      <td>0.177230</td>\n",
       "      <td>0.177</td>\n",
       "      <td>...</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.267865</td>\n",
       "      <td>0.215347</td>\n",
       "      <td>0.279739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EVM prediction%2hcontig_0000_003.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.127193</td>\n",
       "      <td>0.245614</td>\n",
       "      <td>0.134</td>\n",
       "      <td>...</td>\n",
       "      <td>TE gene</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene_ID          GS          HE        IT0  \\\n",
       "0        gene_model_hcontig_0000_03.1    0.487427    0.000000   0.000000   \n",
       "1        gene_model_hcontig_0000_03.2    0.577405  144.208849   0.000000   \n",
       "2        gene_model_hcontig_0000_03.3  564.783136   52.086774  24.713948   \n",
       "3        gene_model_hcontig_0000_03.4    0.000000    0.167865   0.000000   \n",
       "4  EVM prediction%2hcontig_0000_003.5    0.000000    0.000000   0.000000   \n",
       "\n",
       "           IT6         IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0     0.000000    2.377196    0.384531          0.239374          0.147651   \n",
       "1  1016.181477  456.791691    0.446980          0.190050          0.110448   \n",
       "2   235.779282  239.133947  316.475087          0.172863          0.085502   \n",
       "3     0.115347    0.179739    0.000000          0.224765          0.177230   \n",
       "4     0.000000    0.000000    0.000000          0.127193          0.245614   \n",
       "\n",
       "   5mC_upstream_50     ...      Gene_type  Top_500_5mC_gene_body_50  \\\n",
       "0            0.213     ...       Effector                      High   \n",
       "1            0.176     ...       Effector                      High   \n",
       "2            0.148     ...      All genes                      High   \n",
       "3            0.177     ...      All genes                      High   \n",
       "4            0.134     ...        TE gene                      High   \n",
       "\n",
       "   Top_500_5mC_upstream_50  Top_500_5mC_downstream_50  Top_500_5mC_both_50  \\\n",
       "0                     High                       High                 High   \n",
       "1                     High                       High                 High   \n",
       "2                     High                       High                 High   \n",
       "3                     High                       High                 High   \n",
       "4                     High                       High                 High   \n",
       "\n",
       "   Top_500_6mA_tss_50       UG_01       HE_01       IT6_01      IT9_01  \n",
       "0                High    0.484531    0.100000     0.100000    2.477196  \n",
       "1                High    0.546980  144.308849  1016.281477  456.891691  \n",
       "2                High  316.575087   52.186774   235.879282  239.233947  \n",
       "3                High    0.100000    0.267865     0.215347    0.279739  \n",
       "4                High    0.100000    0.100000     0.100000    0.100000  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "h_contig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Top_500_5mC_gene_body_50</th>\n",
       "      <th>Top_500_5mC_upstream_50</th>\n",
       "      <th>Top_500_5mC_downstream_50</th>\n",
       "      <th>Top_500_5mC_both_50</th>\n",
       "      <th>Top_500_6mA_tss_50</th>\n",
       "      <th>UG_01</th>\n",
       "      <th>HE_01</th>\n",
       "      <th>IT6_01</th>\n",
       "      <th>IT9_01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_pcontig_000.1</td>\n",
       "      <td>9.272092</td>\n",
       "      <td>0.061132</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.693499</td>\n",
       "      <td>2.731541</td>\n",
       "      <td>0.232882</td>\n",
       "      <td>0.129151</td>\n",
       "      <td>0.251958</td>\n",
       "      <td>...</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.484531</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>2.477196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_pcontig_000.2</td>\n",
       "      <td>0.392567</td>\n",
       "      <td>0.198648</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.925084</td>\n",
       "      <td>0.383660</td>\n",
       "      <td>0.197059</td>\n",
       "      <td>0.120588</td>\n",
       "      <td>0.193000</td>\n",
       "      <td>...</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.546980</td>\n",
       "      <td>144.308849</td>\n",
       "      <td>1016.281477</td>\n",
       "      <td>456.891691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_pcontig_000.3</td>\n",
       "      <td>0.516801</td>\n",
       "      <td>130.454256</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>950.078729</td>\n",
       "      <td>419.259148</td>\n",
       "      <td>0.395637</td>\n",
       "      <td>0.198405</td>\n",
       "      <td>0.153539</td>\n",
       "      <td>0.171000</td>\n",
       "      <td>...</td>\n",
       "      <td>Effector</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>316.575087</td>\n",
       "      <td>52.186774</td>\n",
       "      <td>235.879282</td>\n",
       "      <td>239.233947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_pcontig_000.4</td>\n",
       "      <td>504.229774</td>\n",
       "      <td>46.756651</td>\n",
       "      <td>21.959973</td>\n",
       "      <td>217.808881</td>\n",
       "      <td>219.104632</td>\n",
       "      <td>279.622402</td>\n",
       "      <td>0.233271</td>\n",
       "      <td>0.091078</td>\n",
       "      <td>0.184000</td>\n",
       "      <td>...</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.267865</td>\n",
       "      <td>0.215347</td>\n",
       "      <td>0.279739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_model_pcontig_000.5</td>\n",
       "      <td>154.614015</td>\n",
       "      <td>32.043647</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>114.040510</td>\n",
       "      <td>450.072586</td>\n",
       "      <td>197.258316</td>\n",
       "      <td>0.232236</td>\n",
       "      <td>0.082611</td>\n",
       "      <td>0.196000</td>\n",
       "      <td>...</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    gene_ID          GS          HE        IT0         IT6  \\\n",
       "0  gene_model_pcontig_000.1    9.272092    0.061132   0.000000    0.000000   \n",
       "1  gene_model_pcontig_000.2    0.392567    0.198648   0.000000    0.000000   \n",
       "2  gene_model_pcontig_000.3    0.516801  130.454256   0.000000  950.078729   \n",
       "3  gene_model_pcontig_000.4  504.229774   46.756651  21.959973  217.808881   \n",
       "4  gene_model_pcontig_000.5  154.614015   32.043647   0.000000  114.040510   \n",
       "\n",
       "          IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0    0.693499    2.731541          0.232882          0.129151   \n",
       "1    1.925084    0.383660          0.197059          0.120588   \n",
       "2  419.259148    0.395637          0.198405          0.153539   \n",
       "3  219.104632  279.622402          0.233271          0.091078   \n",
       "4  450.072586  197.258316          0.232236          0.082611   \n",
       "\n",
       "   5mC_upstream_50     ...      Gene_type  Top_500_5mC_gene_body_50  \\\n",
       "0         0.251958     ...      All genes                      High   \n",
       "1         0.193000     ...      All genes                      High   \n",
       "2         0.171000     ...       Effector                      High   \n",
       "3         0.184000     ...      All genes                      High   \n",
       "4         0.196000     ...      All genes                      High   \n",
       "\n",
       "   Top_500_5mC_upstream_50  Top_500_5mC_downstream_50  Top_500_5mC_both_50  \\\n",
       "0                     High                       High                 High   \n",
       "1                     High                       High                 High   \n",
       "2                     High                       High                 High   \n",
       "3                     High                       High                 High   \n",
       "4                     High                       High                 High   \n",
       "\n",
       "   Top_500_6mA_tss_50       UG_01       HE_01       IT6_01      IT9_01  \n",
       "0                High    0.484531    0.100000     0.100000    2.477196  \n",
       "1                High    0.546980  144.308849  1016.281477  456.891691  \n",
       "2                High  316.575087   52.186774   235.879282  239.233947  \n",
       "3                High    0.100000    0.267865     0.215347    0.279739  \n",
       "4                High    0.100000    0.100000     0.100000    0.100000  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p_contig_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAW0AAAHPCAYAAACY1YvHAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMS4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvNQv5yAAAIABJREFUeJzs3XucVVX9//HXewZFkkQd0ARFNLQ0TdPxlqaikKOJfX9lpmWMmpqWl0izzL5FZllmfRPtm+LlC3wzTc0KjERQ0Sw0kcoLapLfUUETHBVDUBn4/P5Y68CewzlnZpjZ+9w+z8fjPOasvffZa53b56xZe11kZjjnnKsODeUugHPOue7zoO2cc1XEg7ZzzlURD9rOOVdFPGg751wV8aDtnHNVxIN2GUiaLOmOcpcDQNLHJT0jqUPS5HKXp15JmiDp8S6OGSHJJDX3QX6HxnMN7u25enreDcm7O69PvajYoB0DmxW4PVjusvWBc4ETy12I6Drg18D2hHKtp8h78WDeMf0lXSnpFUlvSpomadu8Y4ZLmh73vyJpoqSNU3tm1eVy4JBcoi9/2CXNkXRVX5yrj/wZ2AZoL3dBkiqpMlVKv3IXoAuzgc/lbXsnzQwlbWxmqeZhZsvSPH93SdocGAzMNLPFXRye/17kv0Y/BT4OnED4Mv4EuEPS3ma2WlIj8Pu47yNAEzAFEHB2b59LtTOz5cDycpcjC/H79a9yl6NqmVlF3oDJwB0l9h8CrAIOTWw7A3gD2DGm5wBXA1cAr8Xbj4CGxGPagAnADcDrwK1x+zDg5sTjfg/slHjcdsDvgFeBFcBTwPGJ/d8CngPeJnxApxZ7bkB/QtB7GXgLeBA4KLH/UMCAw4GHYn7zgL26eA23IATG14CVhMD7gbxzJm+HFjlPV+/FIEIQ/2ze67MGOCKmj4zp7RLHnBif72Ylzr0zcF887mngKEJwOylxTFfv1QTgceB44J/Av4HfAoPz8joZWBDz+gcwPu+z8oW4/S1gKTAT6Fek3L8Cfp5Ify++xvslti3KvWa5Mibur/feACPi/U8Cs+LnYAEwpovvUf65RnT3MwV8OL7+K4DFwM+7eL+6PG/imMGJbacAz8fjpwNfBCyr97DYa16O2NfVrewF6OLDVjRQxGO+D7wAbAm8H3gTaE3snxPf3Cvj/uOAZcBXEse0EQL9BcBIYCfgXfGNnQx8MD72OkIQfld83PT4xdkD2AFoAVrivk/Gc34MGA40A2cVe26EH5WX4vG7ANcSAtM2eR/yvwCjYnlmAk8CKvH6/I7wY3IwsDswLb5eA4CNgV3jeT8BvAfYuMR78TqwJL4u1wJbJfYfFs8zJO9xTwDfifcvBp7I2z8kPm5UkXwb4jnuBvYEDiAEglXEoN3N92pCfD1/E485IO6/JpHXafE9ODa+n2MJP7Znxf3NQAfwWUJT0h6EgFAsaJ8JPJVI/4kQJL4e0zvF5z4sUcZc0B5ICPqz4vvynvh+jYiPeSqWbyfCj3I7MLBIOQYRmiNuSJyrsTufqfiZWQ6cF/PaD5gL3FbiM9ed8+aOGRzTBxB+0L9G+JE+Lb5Wljhvqu9hsde83HGw4Gtc7gKUePMnxxd4ed7th4ljNgIeBm4H5gO/yjvHHMIXWolt3wQWJdJtwPS8x50CPJP3uMb45Tguph8Fvl2k7F8h1Ao3KvHc7oj3NyXUUsfl5fVP4JK8D/kRiWMOjNu2LZJHLigcnNg2iPCjdWpMD6YbNQpC7eYYwpd4LPB3Qq2nf9z/mfheKe9x9+S+VMAk4J68/YqPO6FIvkfE/cMS2z4cy3xSD96rCYSa1aDEMRcBCxPp54HP5eX/ZWBBvP+J+Nq9u5uf311iObch/LC8DXyd0BQFIcA8kzh+AjFo539GEttGxHN+IbFtWNx2UImyzAGuytvW5WcKmApcn/e4PeMxWxXJqzvnzR2TC9o3AXfmnWcS6wftVN/DQq95Jd4qvU37fuD0vG2v5+6Y2SpJnyHUxpYQanz5HrT4jkRzge9K2szM3ojb5uU9Zm/CL/W/JSW3vwt4b7x/BXC1pBZCTfA3ZvZI3Hcr4aLe/0maCdwJTDOztwuU772EH58/JZ7XaklzCTXhpEcT91+Mf7ci/JudbxdC7WVu4rzLJD1W4LwlmdnNieRjkh4h1HI+RvjBLEaEL+faUxXLosj29wMvWuf29ocJzyunO+8VwHPW+VrCi4TXDklDCM0510j6eeKYfvE5QKiBPce69/Qu4HYz+3fBJ2T2pKSXCQHqFcKP8M3ANyVtFLfPKfK8u1Lsc9CX51pEeG1HSvp04pjc6/FewnduQ86b7/2E/1yTHiL8sCVl+h5WqkoP2ivMbGEXx+xP+Dd6c8K/26+XPrygN/PSDcDfCDXMfK8CmNn18Y0/ChgN/FnSpWY2wcxekPQ+QrveaODHwLcl7Wdm+XnlPlCFAlf+tlUF9hXrAaQi24vl1W1m9qKkRYTaPIR/QRsJNfeliUO3Ivzw5o45MO9Ug+PjXi6SVX7QL6TL9ypalbfPWPfa5f6eQWhKWI+Z/VvSXoSmpjHAhcD3Je1jZi8WegyhLXgU4TW518zaJL0C7EO4JvO1Uk+shLXPxcws/lhtaE+wUp+pBkJT038VeFxXF657+lntzmeyHO9hxanYLn/dIWkEcBXwJcKv6I2S8n+I9lPnKtj+hNrbGxQ3n9C+/YqZLcy7rQ0EZrbIzCaZ2XGEC4+nJ/a9ZWa/N7PxhC/pB1g/aAEsJDSPHJR4Xo2ENrsFpV+BkhYQ3t8DEufdjNDE0ZvzEvvXDiO0HwI8QvhCjUkcsy2htp/7As0FdsnrBjiG0GzwCIU9CQyTNDSxrZnOn9tuvVelmNnLhCD03gLnWJg4rsPM7jGzCwntqpsCR5c49RxC0D6UdbXq+wifk2GUrmm/Q/hB6wsbeq75hAvX670mZrayj8oG4X3eN29bfrqkPnoP+/I1T02l17T7S3pP3rbVZrY0BrZfAPeZ2TWSbgMeA74N/Gfi+KHATyX9NyFgfRW4pIt8bwTOB34n6VuEtrLtCF3arjazZyRdAfyB0Ga+GeFC5AIASScRXtuHCO3wnyYEtWfyMzKzN+O/cz+ItbD/I1wc2Rr47y7KWVQs4+8I/y6eTvgP5HuEC6S/7O55JA0ktCf+mhCkRwCXEv41/k3Ma5mk64EfSVrCui5/jxJ6rED4V/QJYKqk8whd/n4EXFviB3QW4drAFEnnEy6g/oTQzp2rmXX5XnXzqU4ArpT0OjCD0GS1F6E9/VJJRxOaBO4n1OBHAe8mBJxi5hDewxGsC9BzCBdyF1rpbpZtwJHxP7Z2QlvshmoD9o2VnOV0/g+klB8CD0q6GriGcFH//cBYM/tCL8qTbyLwgKSvEnqEHAz8vw04zwR69x62kfeam1l+7b78yt2oXuxG4a5KRryISAjM/yLRY4FQc1tFvCjDui5/VxGC1muEporGxGPagPML5L818D+E4PQ2IZjewLqLJ1cSgnCu69DNrOsJ8B+EmuXrhKaXh4Gj855bsS5/b1O8y1+yi9SIuK25xGtYtMtf3N/lhUhCoJwZX4d3CG2Ck0l03YvHbRJfk3bWddvKP2Y4cEfc3x6P79/F52BnwpfsbUIAPzqW49M9eK8mkLjIF7edBCzP23YCoXb5VnzNHiB24yT8J3RvLPdKwoXYk7vxOX6JzhcYc+/btXnHTcg7bgjhh+7fufeo2Hsetx3bxWs4N77uRucufyU/U4T/bO4k/Ni/SagYXVwiry7PW+SYUwg9m1bGz855wMpir08a72Gh1zyN2NbbW64LTk2SNIfwRp9V7rK4viFpD0IbdrOtu/Draoyk/wJGm9nu5S5Lpan05hFX5yT9P0IN7xlCje0nhC6H88tYLNfHYtPILELzzWjCBcVvlLVQFcqDtqt07ya0rW5H+Hd3DjDeavlfxPrUTLg2MYjQvHUhoVuty1PTzSPOOVdrqrrLn3PO1ZuqDNqSTpJUczOi9cUUmn0173C1v8aSBscpZA8tcUyqU3FWy1SfORsyz3WR85ikY/uqXD3M+1hJNd18kHrQltQW+9j2pV8BO/bxOTPTV1+OvlLkS1bVr7ErrZLm2O6rika9qMoLkRZGY/XliCyXx19j5ypTlzVtBecpLEn1tqRFki6N++7J/7WWtJmkFZI+EftJb08YKWfJf1vi/sfiOV+QdFFyuHnc/6iklZJelXSfpK3jvk7/uud+qSUdL+mfkv4t6bf5NVlJJ0taIOktSf+QNF5S0dcgcd7W+B/Dckn/I2ljSV+M5W6X9JPkeeL+H8bX6k1JD0s6Iu4bQejgD7A0vi6Tk++JpO8rrOyyRNLluXNL+lahGomkP0maWOQ57CPprni+NyQ9ICk5tL0t3r01lqWt0Gsct31B0kJJ78S/p+XtN0mnS7o1Pu9nJZVcoaer8nX3vPE8j8T39q+EaUR7RGH1nZ9Kejme50FJyekFHpL0tUT6xli298T0u+Jrs950BZLGxc9K/7ztN0qaVqJMJulMSb+L36t/SBolaVtJM+Pr8TeFOTWSj/tw/M6skLRY0s8VpjEgft4OAb6kdSsRjUg8fI/4XFdImpc7t6RN43t0bF5eYyStUvx+FngOP5D0tMJ3uU3SZZI2iftOIoxi/kCiLCfFfYMkTYrfg3/H59Ocd+5xkp6LZb2DMNCqJEk7x3O9Fct1lMJ3+6TEMcMk3SzptXj7vaSdEvv7JOYofKf+Efcvje9p6cp0N0Z0XUoY2XcKYY6HA4AvJkYfvUpiVBthkvElhCGkWxJGOX2HOEdtPGZvYHXcvjNhftvlwNlx/3sIo97OI/TN3Q04Fdi60Ego+mCu3SLPPXfe22MZjojpPxBG4O1CGG67Cvhk4nE3EkY1HkxoYjgrPp89CHMbfIIw4mrX+FwHxcfNIQxXvji+LseRmLoU2Dam903k9b54rj2KjKw7jLDizC6EIchXEbrO5UYL5ua0PjWWZUiR1zj3PM+KZTs7psfmjcxbRFjcYCThs/MOsH2J17hk+bpzXsL8EUsIsyvm3qcn6Xq052R6Nq/5D0lMIRrLtJR1I+7GEPqUb5R/fsLI0teI08XGbYMIoxQ/XqKMRphT4wTCBF2/JHxuZxKG6u9MGLL9aOIxJefBpndzbF8DzMgr402EWS4LjtAkjF4+kPBdPoow1cB3E6/L5YQ5wnNlGUCYROoBwoIW+8b3/buEkZm592M/woyPF8XX4QuE0Y5W4vWs2jna1563i4A9kDAc9Iwi+/sTpp1MrtjyEHB5It1G3jBxQlDLn1t5AuuGqO8V3/iCX3YKB+1ezbVbImivzDvvbYQv6saJbXOI8xUT5jZYAwzPO9dvgf+O9w8lbxhv4jxz87bNAq5LpO8gzKmRS/8QmJdX5sdLPCfFD9KJxb5kRV7jPwE35B0zGXgg7zyXJtL9CEHpxGLl6UH5ip6XMAHT6yQWAiAE+G4Hbbo3r/mRhC9qP0Iw/DdhPpfcnOHfA2YVOn9MX0XnoH8m4Utc9Eta4LnvFrclF/Lo9HmiG/Ngs+FzbOcCTW7Khi0I35Gj88pcalj9GXT+bq73mSX8mC8HBuRt/xtwQbz/y+TrHbddR+mgXbVztOduXTWP7EoIzHcX2mlhfuj/jU8SSbsSfhVv6OK8u5CYPzp6gDCj22aEEW+zgccl/Tr+eziki3N2d67d5bkb8AM6z7lcyPN5530Z+Id1XkfyZdbNZ7wXIfAsyMvrY93ICzrPQ9zpeUTXAsdLGqAwadbngOuLnUzSVpKuif+CLSMEmq0I84D0RLH3rOic32bWQfiBKzrXcw/KV+q8uxBqmsnmnLn0TMF5zeN5cs/xj4Tvwz6EAPdHwuf00Lj/UErP3HctMEbrZjo8BZgSn08pyc9Ebhrbxwpsy70eewMn5n3+cs+rp5/BTvN1m9m8mHdr3P4Zwn8Qfyh2MoUeHQ9I+lcsy3/R9edvb0KNd2ne89gt8Rx2Yf33uav3vadztOfyXUb4gdqQOdqLxZzk/N43KjTDvruL8nd5IbLUnMw51wGPShoOfJ5QU+xq6s9S8+eahUUAPkqYRvWj8byXSjrEzP5e5HG9mmu3hELnLbQtN6VjQ0zvU+C47lzYK/U8IPy7uIKwpNkywjziN5U43xRCO994wn89bxN+hDdkFfRC71n+tq7Kv6HlK3Xe7nxOu9LlvOZmtlzSfELTwQcI1ybmAtvH9s59CMvWFWRmf4+PP0nSbwm11pJt/lGhuanTmge7q3MTz/1lwnJ/pwCT4w/ceiTtT5hM7TuE9/h1wipIl3dRhgbCj9FHCuzLzQq5Ie97tc/R3mXQXkD4Eh1OgWlFY8ZPSMqtMnEi4V+EpEJz1C4gMX90dBCheeTf8bxG+ELMlXQxoR3q04RaeI+Y2cuScnPtTu3p43vor4QPxnvM7N4ix+Rq6T2eu9fMOuKFpFMIQft2Myu18MNBwDlm9nuAeLFom7xjVnWjLE/GcyX/izqIXs7N3c3ydWUB0CppU1u3yMT+PTxHcl7zZ2NZcvOaJ6eynUMI2rsAPzWzt+Ln/yLC6/iXLvK5lhDYBwN/MrOne1jO7lg7D3aJY3ozd/QvCJ0LziL8Z1kouOUcCCw2s+/mNkjavhtlmU/4MV9jZs8WOfcC1n+fu3rf187RngiMheZoP4EwR/uGLKrS7ZgT/8u6B7hH0rcJ12aOJiy3VlDJ5pEYQK8g1HJPlvReSftKOjPv0NwHcVNC/96kNuAj8Wps7srqj4FD4hXYnSV9lnDR5DIIv86SvqnQI2A44Zd5O3oXICYAF8Srt++TtFu88nxhL865HjP7B6HNfnL8t3BHSc2Szpf0iXjYc4Rf5Y9JGqIwZ3VPXEe4+n80JZpGon8Q/lXeVdI+hFrPO3nHtAGHS3qPpC2KnOdHwOckfUnSTpLOJlxAuayHZd+Q8nXll4R2yhskfUDSGNavPJQUg31uXvOjJO0S0/nzms9h3TzM8xPbTgT+bF3Pv3wT4WLbmXT93m2oHxLmz75a0ockjZR0tKRrEse0xWNGKAxE6vaYjdgkcCvhe3y/lZ6z/B+EIPnZ+F04kxAQk9oI/63sFcvSn9Ds9CfCPOlHStpB0gGSviMpV/ueCIyWdGH8TJ5G1/NwJ+do3yP+J1BojvaXY96HxLwPlvRjJXqQdMMESsSc+J6cG9+j7QlNTV3N0d6t3iMNhAVJnyV8mV4Avpd3zLsIbZE3FHj8/oTa8Vt0XqjzE4S2sdw5L2LdFepdCG1kufmlFxIvPsT9J7H+hchezbVb5LkXOu9VwJy8bTeTWKGa0DY6IfGa/YuwEvreiWP+k3DBbQ3h30sofHFoMgUWGyX8Ov8T1ltMt1OZCVekHyI0zfyT0Ab+ODAhccxYwn9Sq4C2Eq/fGfG9WBX/npa3v9AFzTYKzFfew/J1eV5CT4L58fPy9/icun0hMqZLzmsejxkYn3/ycYfGvC7q5nt3A+H7smk3vn+dnjsF5kAntNMasFtiW8l5sOnFHNtx+8Fx+7hulPlSwjWIXE+sM+kcC/oTLvC/RucLgu8mVBoXsS5O3EyoveYeezLhgt9KQsw4K3nuIq9pVc/R3icTRiksB/U8cIiZ5V+scimQtAC40cy+V+6yuJ6R9AdCU2D+wrVVQ2Gx32uAoWa2otzl6Q1V2RztvRoRqbCq9DaErk5/9YCdPklbEX69RxC+NK5KSNqSMFf0Rwn/YVQdSe8ifPa+QVh9p+oCtqp8jvbeDmM/kFC9f4YwEMSl72VC3/gvmNkr5S6M65H5hAFn3zCzap1r4wJCU+YDhMEu1aiq52j3+bSdc66KVOXUrM45V688aDvnXBWpyqlZszJ48GAbMWJEuYvhXN155JFHXjGzrqauqEsetEsYMWIE8+bNK3cxnKs7kp4rdxkqlTePOOdcFfGg7ZxzVcSDtnPOVREP2s45V0U8aBcgaaykScuWLev6YOecy5AH7QLMbLqZnT5o0KByF8U55zrxoO2cc1XEg7ZzzlURD9rOOVdFPGg751wV8aDtnHNVxIO2c85VEQ/adWrevHkcdthhPPJIxS+J55xL8KBdpyZMmMCaNWv49re/Xe6iOOd6wIN2HZo3bx7Lly8HYPny5V7bdq6KeNAuoNaHsU+YMKFT2mvbzlUPD9oF1Pow9lwtu1jaOVe5PGjXoYEDB5ZMO+cqlwftOpTfPPKd73ynPAVxzvWYB+061NzcvLZ2PXDgQPbee+8yl8g5112ZBG1JO0gaJekoSftI2iSLfF1xEyZMoKGhwWvZzlWZ1FZjlzQCOBM4ARgGKLH7HUl/BCYBvzazNWmVwxW2+eabM2DAAGr1YqtztSqVmrakK4C/AzsCFwG7AoOAjYH3AEcBDwDfBR6VtE8a5XDFXXLJJbz55ptccskl5S6Kc64H0qppvwO818xeKbBvCXBPvH1H0lHA9sDDKZXF5Vm4cCFtbW0AtLW1sXDhQkaOHFneQjnnuiWVmraZfbVIwC507Awzuy2NcrjC8mvXXtt2rnp475E6lKtlF0s75ypXmhcip3XnODM7Jq0yuML69etHR0dHp7Rzrjqk+W09GngOmJNiHm4DJAN2obRzrnKlGbQvB04EDgb+B5hsZotSzM91kyTMrFPaOVcdUmvTNrMLgO2A8UAz8IykP0g6VtJGaeXrunbIIYeUTDvnKleqFyLNbLWZTTOz/wB2AO4FLgEWS/JZisrkxBNPLJl2zlWuLHuPbApsDgwElgNW+vC+J2lTSY9IOjrrvCvJLbfc0il96623lqkkzrmeSjVoSxogqVXS/cBjhEE0rWa2o5m92Qfnv0HSEkmP521vkfS0pIWSvp7Y9TWgc8SqQ3fffXen9OzZs8tUEudcT6XZ5W8S8GngGeB64Bgze72Ps5kMXAVMTeTbCPwMGAMsAh6O3Q+HAguAup+savXq1SXTzrnKlWbvkVOB54GXgCOBIwv1UuhNP20zuz9OTJW0L7DQzJ4FkHQz8HFCs8ymhHlQVkqa4RNVOeeqTZpBeyplaLcmzCj4QiK9CNjPzM4CkHQS8EqxgC3pdOB0gOHDh6dbUuec66HUgraZndTVMSnNq12o0/HaHw8zm1zqwWY2iTBlLM3NzeX40UndNttsw0svvbQ2PXTo0DKWxjnXE6ldiJR0Whf7NwHuSCHrRYT+4TnbAi+mkE/VGjFiRKf09ttvX56COOd6LM3eI1dJ+kShHZL6A9OBNOYDfRjYKa6WszFwPNCteVAS5RsradKyZctSKF75PfTQQyXTzrnKlWbQPgv4haRRyY1xNOTvgF2Aw3qTgaSbgLnA+yQtkvR5M+uIec8EngRuMbMnenJeM5tuZqfX6qoua9asKZl2zlWuNNu0r5U0GPitpFFmNj/WfH8DfBA4ONfDoxd5nFBk+wxgRm/O7ZxzlSjtYeyXAtcBd0raDbgN2Bs4zMwWppl3b9R684hzrnqlPozdzM4Dfg/MB/YjBOyn0s63N2q9ecQ5V73SHBE5MZF8C1hDGMp+RnKQjZmdk1YZXGFbbbUVS5YsWZveeuuty1ga51xPpDm4Zve89FygMW97TfaDrnSvvNJ5+c6lS5eWqSTOuZ5K80LkqK6PqkySxgJja3WF8oaGhk49RhoafKlQ56pF2rP8vVvSGElHxZ4kVaHW27QPOuigkmnnXOVKs037g8AfgG3ipjckHWtmPg9omfXv379k2jlXudKsaf+AMMvfQYTlxu4jTKPqyuz+++8vmXbOVa40g3YzcLaZ/dnM5gOnEIaXV/wyY7XeT3vw4MEl0865ypVm0B5MqGkDYGbtwApgSIp59olab9NevHhxybRzrnKl2eUPYAtJHYm0xW1rq7Bm9mrKZXB5fO4R56pX2kF7QV5ahFn4cveN0HfbZUgSZtYp7ZyrDmkG7artp+2cc5UqzcE196V1btc7yVp2obRzrnKVbSicpGZJd5Yr/1JqvfdIv379Sqadc5Ur7RGRYyT9SNL3Je0Yt+0s6XdAxS6XUuu9R84+++xO6XPPPbdMJXHO9VSaa0S2ElaPORn4OjBX0vGEC5GvAnuYWUta+bvi/vrXv3ZKP/LII2UqiXOup9KsaY8HvmFmgwnrNA4BvgrsZWYnm9njKebtSpgzZ07JtHOucqUZtN8L/Crevw1YDXzFzP6ZYp7OOVfT0gzamwJvApjZGsJCCC+kmJ9zztW8tLsNfCwx+rEBOELSy8kDzOz2lMvQY7U+n7ZzrnqlHbSvz0v/LC9dkSMizWw6ML25ufm0cpclDT4i0rnqlVrziJk1dONWcQG7HhxyyCEl0865yuXrTNWh/H7a+WnnXOVKJWhL6vb6VZIGSspfBNg551wBadW0r5d0t6QTJG1W6ABJH5R0GbAQ2COlcrgCJk2aVDLtnKtcaQXtDwC3A98CXpX0lKR7Jf1B0oOSXiMMYx8GHGZmv0ipHK6AWbNmlUw75ypXKr1HzKyD0FPkZ5KaCetEbg8MAB4BfgTc6wsglIcvguBc9Up9ejczmwfMSzufvuT9tJ1zlcp7jxRQ67P8Oeeqlwdt55yrIh6061BDQ0PJtHOucvm3tQ4NGzasZNo5V7k8aNehpUuXlkw75ypXJkFb0pGS7pC0QNJ2cdupkg7PIn/X2RZbbFEy7ZyrXKkHbUmfBW4BngF2ADaKuxqBC9LO363vpZdeKpl2zlWuLGraFwCnmdl4oCOx/UFgzwzyd865mpFF0N4JmFtg+3Kg4LwkLl3ee8S56pXFt/VFYOcC2w8GfL3IMvBh7M5VryyC9iRgoqQDY3o7Sa3AZcDPM8jfOedqRhZzj1wmaRAwC9gEuBd4G7jczPKXH6sIPveIc65SZdKYaWYXAYOBfYH9gSFm9p9Z5L0hfO4R51ylSr2mnWNmK6iy2f6cc67SpB60Jd1LWHU9nwFvEVaumWJm89Mui3POVbssmkeeBPYCtgEWxds2cdsSwgIJD/noSOec61oWzSNvAZPN7MvJjZJ+DJiZ7S3pCuAS4O4MyuOcc1Uri5p2K2HpsXzXACfH+5OAXTMoi3POVbUsgrYIC/3m2zXuA1gF+AiPjPiISOeqVxbNI1OA6yXtBDxMuAC5L/A1YHI85hDg8QzK4oDddtuNRx99dG169913L2MUO4IyAAAgAElEQVRpnHM9kUXQPh94GRgPvCdu+xdhRfbLY3om8IcMyuKAxx/v/Pv42GOPlakkzrmeymJE5GrgB8APJG0Wt72Rd8zzaZfDreNzjzhXvTIbXAPrB2vnnHM9k0nQlnQycAIwHNg4uc/MdsyiDM45VwuyWLnmq8CPgUeAEcBvCRcdtwRuSDt/55yrJVn09ToNON3MLiR07bvKzI4hBPLtM8jfOedqRhZBe1vgL/H+StatVnMT8MkM8gdA0i6SrpZ0m6Qzs8rXOef6UhZB+1+EaVkBngMOiPdHUngiqW6TdIOkJZIez9veIulpSQslfR3AzJ40szOA44Dm3uTrnHPlkkXQvgc4Jt6/HvhJnPnvV8DtvTz3ZKAluUFSI2HY/JGEUZcnSNo17jsGeACf48Q5V6Wy6D1yOvHHwcyulvQacCDwa8L8IxvMzO6XNCJv877AQjN7FkDSzcDHgQVmNg2YJun3wC97k3c123LLLXn11VfXppuamspYGudcT2QRtLcFXsglzOxXwK8kCdgO6OuBNcOS+RGmgt1P0qHAJ4D+wIxiD5Z0OuGHhuHDh/dx0SpDMmADtLe3l6kkzrmeyiJo/x9h/uwledu3jPsa+zg/FdhmZjYHmNPVg81sEmHWQZqbm3vV5u6cc30tq1n+CgW/gYS5tvvaIkINPmdb4MUU8nHOucylVtOWNDHeNeBSSSsSuxsJbc9/SyHrh4GdJO0ALAaOBz7TkxP4auzOuUqVZk1793gTsEsivTuhu9984KTeZCDpJmAu8D5JiyR93sw6gLMIMwc+CdxiZk/05Ly+GrtzrlKlVtM2s1EAkv4HODeNyaLM7IQi22dQ4mKjc85Vq9TbtM3s5GTAljRA0mhJPoTdOed6KIsJoyZL+mK8vzFhSPtdwNOSjkw7/w0haaykScuWLSt3UZxzrpMseo8cATwY7x8DvJuwgs2EeKs43qbtnKtUWQTtLVjXR7sF+LWZLQFuxldgd865Hslqwqjd4pwgRwCz4/aBhKlaK443jzjnKlUWQfsGwuRQjwOrWTdZ037AUxnk32PePOKcq1RZLOx7saQnCEuN3Wpm78RdHcAP087fOedqSRY1bczs12b2X2a2KLFtipn9Lov8nas17e3tnHPOOT7ZVx3KJGg75/rWlClTeOyxx5g6dWq5i+Iy5kG7AL8Q6SpZe3s7d955J2bGnXfe6bXtOuNBuwC/EOkq2ZQpU1izZg0Aq1ev9tp2nfGg7VyVmT17Nh0dHQB0dHQwa9asMpfIZSmzoC2pWdKnJW0a05tKymIRBudqyujRo+nXL3x1+vXrx5gxY8pcIpelLOYe2VrSQ4Q5R34JbB13/QT4cdr5O1drWltbaWgIX93GxkbGjRtX5hK5LGVR0/4vwqjIJiC5EMKtwEczyL/H/EKkq2RNTU20tLQgiZaWFl+Yuc5kEbQPBy4ys9fytv+TMOCm4viFSFfpWltb2X333b2WXYeyaFMeALxTYPsQ0lkj0rma19TUxMSJE7s+0NWcLGra99N5WTGLk0d9jXXzkDjnnOuGLGraFwD3SdoH6E+4+PgBYBBwYAb5O+dczchiubEFhMV8/0xYsWYTwkXID5nZP9PO3znnakmqNW1JGwHfA35mZt9OM6++JGksMHbkyJHlLopzznWSak3bzFYBXwSUZj59zXuPOOcqVRYXImcCh2WQj3PO1bwsLkTeDXxf0geBR4A3kzvN7PYMyuCcczUhi6B9Vfx7ToF9BjRmUAbnnKsJWSw35jMJOudcH/GA6pxzVSSToC3pY5Lul/SKpKWS7pN0VBZ5O+dcLcliatZTgd8QJoj6GvB14P+A30g6Je38N4TP8uecq1RZ1LS/BnzFzE42s+vj7STgfEIArzjeT9s5V6myCNrDgTsLbP8DsH0G+TvnXM3IImg/DxRaD+mjwHMZ5O+cczUji37alwNXStqLMGmUAQcBnwPOziB/55yrGVn0075G0hLgPOATcfOTwHFm9ru083fOuVqSyWroZvYbQg8S55xzvZBJ0AaQdBiwa0wuMLN7ssrbOedqRepBW9IOwO2EhRBejJuHSnoM+KSZPZt2GZxzrlZk0XvkeuANYEczG25mw4EdgdeB6zLI3znnakYWzSMHAPub2fO5DWb2vKTxwNwM8nfOuZqRVT/tAQW2bwK8kEH+zjlXM7II2ucBEyXtL6kx3vYHfhr3VRyfe8RVuvb2ds455xza29vLXRSXsSyC9k3AnsCfgLfi7U/AXsCNkt7I3TIoS7f43COu0k2ZMoXHHnuMqVOnlrsoLmNZtGmflUEeztWN9vZ27rzzTsyMO++8k3HjxtHU1FTuYrmMZDEickraeThXT6ZMmcKaNWsAWL16NVOnTmX8+PFlLpXLShbzae8q6X2J9BhJv5B0oSRfH9K5Hpo9ezYdHR0AdHR0MGvWrDKXyGUpq37aHwKQtC3wO2BL4EvAJRnk71xNGT16NP36hX+S+/Xrx5gxhSbRdLUqi6C9CzA/3v8U8JCZHUWY5e+EDPJ3rqa0traubR5Zs2YN48aNK3OJXJayCNqNwDvx/uHAjHj/n8DWGeTvnHM1I4ug/ThwpqSPEIJ2bhWbYcArGeTvXE2ZMmUKkgCQ5N3+6kxWa0SeBswBbjKzx+L2Y4C/ZJC/czVl9uzZrF69Ggi9R/xCZH1JPWib2f3AEGCwmSVXX78GODPt/J2rNX4hsr5lUdPGzFab2Wt529rMbEkW+TtXS1pbW2loCF/dxsZGvxBZZzIJ2s65vtPU1ERLSwuSaGlp8dGQdSazlWucc32ntbWVtrY2r2XXIQ/azlWhpqYmJk6cWO5iuDLw5hHnnKsiqdS0JR3c3WNj75JMSPoP4GPAVsDPzOyurPJ2zrm+kFbzyBzAAMW0xb/5aQgjJjeYpBuAo4ElZrZbYnsLcEU8/3Vm9gMz+y3wW0lbAJcDHrSdc1UlreaRIYTa7BBCQH0aGAeMjLdxwFOEATa9NRloSW6Iswf+DDgS2BU4QdKuiUO+Gfc7V5V85Zr6lUrQNrP23A34LnCumd1oZs/G243Al+mDWf5i88qreZv3BRbGvN4BbgY+ruCHwB/MbH7+uZyrFr5yTf3K4kLkrsCiAtsXA+9PKc9hdF40eFHcdjYwGjhW0hmFHijpdEnzJM1bunRpSsVzbsPlr1zjte36kkXQfgL4tqS1K7LH+9+K+9KgAtvMzCaa2d5mdoaZXV3ogWY2ycyazax5yJAhKRXPuQ1XaOUaVz+yCNpnAqOAxZLmSJpDqPkeRnpzjywCtkuktwVeTCkv5zLlK9fUtywmjHoY2AH4OmExhL/G+zvEfWl4GNhJ0g6SNgaOB6Z198GSxkqatGzZspSK59yG8wmj6ltWE0atiM0OXzGz8WZ2rZm92RfnlnQTMBd4n6RFkj5vZh2EVeBnAk8Ct5hZt5tizGy6mZ0+aNCgviiic33KJ4yqb5kEbUlHSrpD0gJJ28Vtp0o6vLfnNrMTzGwbM9vIzLY1s+vj9hlmtrOZvdfMvtfbfJyrFD5hVH3LYjX2zwK3AM8Qmkk2irsagQvSzn9DePOIq3Stra3svvvuXsuuQ1nUtC8ATjOz8UBHYvuDwJ4Z5N9j3jziKl1uwiivZdefLIL2ToQ253zLgc0yyN8552pGFkH7RWDnAtsPJqzI7pxzrpuyCNqTgImSDozp7SS1ApcBP88g/x7zNm3nXKXKop/2ZcDtwCxgU+Be4GrgajOryEmbvE3bOVepMlm5xswukvQ9wjwkDcACM1ueRd7OOVdLsujyd4Okd8cBNvPM7C9mtlzSpnEubOecc92URZt2KzCgwPYBhHm1K463abtK5/Np16/UgrakLSU1EWbc2yKmc7fc4ggvp5V/b3ibtqt0Pp92/Uqzpv0KsISwtNgCYGni9i/gOuC/U8zfuZrk82nXtzSD9ijgcEJN+1jCVKy520HAcJ8TxLme8/m061tqvUfM7D4ASTsAL5jZmrTycq6eFJpPe/z48WUulctKFv20nzOzNZKGStpf0sHJW9r5bwi/EOkqmc+nXd+y6PI3NLFazZ+AOYQBNrlbxfELka6S+Xza9S2LLn8/BVYTBtasAD4CfIqwOEFLBvk7V1OampoYNWoUAIceeqjP9FdnshgReQjwMTN7SpIBS83sT5LeBr5LGN7unOsBMyt3EVyZZFHTHkDo/gfwKrBVvL8A+GAG+TtXU9rb25kzZw4Ac+bM8S5/dSaLmvZTwPuBNuBvwBmSXgC+BCzOIP+qduWVV7Jw4cLU8zn33HP77FwjR47k7LPP7rPzuc4Kdfnz3iP1I4ua9hXAe+L9i4GPAs8CXwS+kUH+ztWUQl3+XP1IvaZtZjcm7s+XNIJQ837ezF4p9rhykjQWGDty5MhyFyWVGuuhhx663rYrrriiz/Nx6Rg9ejQzZsygo6PDu/zVoUxWY0+Ks/3Nr9SADd7lz1U27/JX3zIJ2pI+LWmSpN9Kmpa8ZZG/6yx3EatY2lW2pqYmWlpakERLS4t3+aszqTePSPoR8GXCQJoXCRNIOed6obW1lba2Nq9l16Eseo+MA04ws9syyMt10x577AF4W7Zz1SaL5pEGQlc/51wf8fm061dWq7GfmEE+ztUFn0+7vmXRPLI58BlJY4BHgVXJnWZ2TgZlcK5m+OCa+pZFTXtXQvPIO4T+2bsnbrtlkH+P+dSsrpL54Jr6lsXgmlFp59HXzGw6ML25ufm0cpfFuXw+uKa+ZT64xjnXO8nBNQ0NDd7tr86kUtOOg2ZONLM3uhpAY2bHpFEG52pVU1MTQ4cOpa2tjaFDh/rgmjqTVvNIO+sG0filbef6UHt7O4sXhwkyFy9eTHt7uwfuOpJK0Dazkwvdd8713pQpUzpdiPTeI/XF27SdqzKzZs1au3KNmXHXXXeVuUQuS1nMPdKfMHf2KMKqNZ1+KMxs37TL4Fwt2XrrrWlra+uUdvUji8E11wJHA78jLDHmE0Y51wsvvfRSybSrbVkE7WOAj5vZfRnk5VzN69evH2+//XantKsfWbRpL2Hdwr7OuV568803S6ZdbcsiaH8D+L6kLTLIq0/4MHbnXKXKImjfBQwAlkh6QdKzyVsG+feYLzfmKtkhhxzSKV1ozU9Xu7JoDJsKfAD4KfAyfiHSuV4555xzuO++dZeI0lj82VWuLIL2GOAwM3sog7ycq3lNTU0ccsgh3HfffRx66KE+GrLOZNE88jzwdpdHOee67XOf+xybbropJ57o64vUmyxq2uOByyR90cwWZpBf5q688koWLqyup5Yr77nnnlvmkvTMyJEjvTkAmDZtGitWrGD69Ok+hL3OZBG0bwX6A09LehvoSO40s80yKEOqFi5cyN8ef5LV79qy3EXptoZ3wqWFR559ucwl6b7GFa+WuwgVIX+5sXHjxnkTSR3JImiflUEeZbf6XVuy8v1HlbsYNW3AUzPKXYSK4MuN1bdUg7akfsCmwG/N7MU083KuXhRabsyDdv1I9UKkmXUAPwI2SjMf5+rJ6NGj1w5d9+XG6k8WvUceBPbOIB/n6kJyubHGxkZfbqzOZDXL3+WShgOPAJ0mSjCz+RmUwbma0dTUxKhRo5g5c6b3065DWQTtX8a/Pymwz4DGDMrgXE3JLYLg6k8WzSM7lLjtmEH+ztWU9vZ25syZA8CcOXNob/dlWOtJ6kHbzJ4rdUs7f+dqTaEuf65+ZLJGpKQjJd0haYGk7eK2UyUdnkX+ztWSQl3+XP1IPWhL+ixwC/AMoUkk1/2vEbgg7fzzyrKjpOsl3ZZlvs71Je/yV9+yqGlfAJxmZuPpPIT9QWDP3p5c0g2Slkh6PG97i6SnJS2U9HUAM3vWzD7f2zydKyfv8lffsgjaOwFzC2xfDvTFvCOTgZbkBkmNwM+AI4FdgRMk7doHeTlXdk1NTbS0tCCJlpYW7/JXZ7II2i8COxfYfjDwz96e3MzuB/JnEtoXWBhr1u8ANwMf721ezlWK1tZWdt99d69l16EsgvYkYKKkA2N6O0mtwGXAz1PKcxjwQiK9CBgmqUnS1cCHJF1Y6IGSTpc0T9K8pUuXplQ853qnqamJiRMnei27DmXR5e8y4HZgFmHyqHuBq4GrzexnKWWrwkWxdjM7w8zea2aXFinvJDNrNrPmIUOGpFQ853qnvb2dc845x/to16FMuvyZ2UXAYEKzxf7AEDP7zxSzXARsl0hvS2imca4mXHPNNTz66KNMmjSp3EVxGcskaAOY2Qozm2dmfzGz5Sln9zCwk6QdJG0MHA9M6+6DJY2VNGnZsmWpFdC5DdXe3s7s2bMBmDVrlte260wW/bTvlXRPgdvdkn4v6QpJe/Xi/DcReqe8T9IiSZ+PU8KeBcwEngRuMbMnuntOM5tuZqcPGjRoQ4vlXGquueaatSMi16xZ47XtOpNFTftJYC9gG0KzxaJ4fy9gCXAQ8NCGjo40sxPMbBsz28jMtjWz6+P2GWa2c2y//l6fPBPnKsDdd9/dKZ2rdbv6kMUsf28Bk83sy8mNkn5MuDi4t6QrgEuAuwudIGuSxgJjR44cWe6iOLeeXC27WNrVtixq2q2EgS75rgFOjvcnEQbBVARvHnGVrLGxsWTa1bYsgraADxTYvivruuatAry64Fw35CaLKpZ2tS2L5pEpwPWSdiL06jBC17+vEYagAxwCPF7w0c65TgYOHMjy5cs7pV39yKKmfT5hcd/xwD2EwTXj47avxmNmAidkUJZu8S5/rpJ95Stf6ZQ+77zzylQSVw5ZjIhcbWY/MLOhwObA5mY21Mx+aGar4zHPm9mitMvSXd6m7SrZQw891Cn94IMPlqkkrhwyG1wDYGZvmNkbWebpXK3J7+LnXf7qS+pt2pK2BL4HHA5sRd4PhZn1xfSsztWN1atXl0y72pbFhcjrgQ8RuvW9SLgQWdG8n7arZI2NjZ0CtXf5qy9ZBO3DgTFm9lCXR1YIM5sOTG9ubj6t3GVxLt/o0aOZOXNmp7SrH1m0aS8hrFLjnOsDn/rUp0qmXW3LImhfBFwsyTuTOtcHpk3rPGHl9OnTy1QSVw5ZBO1vAh8Flkh6UtKjyVsG+TtXU2bNmtUpfdddd5WpJK4csmjTvi2DPJyrG5tvvjkrV67slHb1I/WgbWbfSTuPvua9R1wle+mll0qmXW3LdHBNtfARkc65SpXF4Jp/U6Jvtg+uca5ntttuO1544YVOaVc/smjTPisvvRFhsM0nCSMlnXM9cO6553L++eevTX/5y18ucbSrNVm0aU8ptF3SfMLAmyvTLoNztaRQ75G99967TKVxWStnm/a9wNgy5u9cVfIJo+pbOYP28cArZcy/KJ9P21UyXyOyvmVxIfIxOl+IFLA1sCVwZtr5bwife8RVMkmYWae0qx/lGFyzBlgKzDGzpzLI37maMnr06E6jIMeMGVPG0ris+eAa56rMcccd1ylo+4RR9cUH1zhXZa677rpO6euvv75MJXHl4EHbuSqTvybk3Llzy1QSVw4etJ1zrop40Hauymy66aYl0662edAuwPtpu0p23nnndUonh7S72pdFlz8k7Ufx1djPyaIMPeH9tF0l+9vf/rZeetSoUWUqjcta6jVtSecDc4GTgD2B3RO33dLO37lakz9sPX8uElfbsqhpnwucY2ZXZZCXczVv9OjRzJgxg46ODvr16+eDa+pMFm3amwEzMsjHubrQ2tq6duh6Q0MD48aNK3OJXJayCNo3AS0Z5ONcXWhqamLYsGEADB06lKampjKXyGUpi+aRF4DvSDoQeBRYldxpZj/JoAzO1Yz29nYWLVoEwOLFi2lvb/fAXUeyCNqnAsuBD8dbkgEetJ3rgSlTptDR0QHAqlWrmDp1KuPHjy9zqVxWUm8eMbMdStx2TDt/52pNcrIogJkzZ5apJK4cfHCNc1WmX79+JdOutqXybkuaCFxoZm/G+0VV4uAa5yrZ8uXLS6ZdbUvrJ3p3wqrrufvFWIl9zrkCtt1227UXInNpVz9SCdpmNqrQ/WohaSwwduTIkeUuinPrGTlyZKeg7Z/T+uJt2gWY2XQzO33QoEHlLopz6/nLX/5SMu1qmwdt56rM6NGjaWxsBKCxsdGHsdcZD9rOVZnW1ta1Qbtfv34+jL3OeNB2rso0NTXR0tKCJFpaWnw0ZJ3xDp7OVaHW1lba2tq8ll2HMgvakoZSeBGE+VmVwbla0dTUxMSJJYdAuBqVetCW9CHgF8D7AeXtNqAx7TI451ytyKKmPYkw099pwIv4gBrnnNtgWVyI3JWwcs2fzazNzJ5L3jLI37ma097ezjnnnEN7e3u5i+IylkXQfgx4Twb5OFc3Jk2axKOPPsqkSZPKXRSXsSyC9jeAyySNlrS1pC2Ttwzyd66mtLe3r13Md9asWV7brjNZBO3ZwL7AXYQ27aXx9kr865zrgUmTJrFmzRoA1qxZ47XtOpPFhciqmzDKuUp29913r5e+8MILy1Qal7XUg7aZ3Zd2Hs7VEzMrmXa1LZNh7LEt+2JJt0m6VdIESVtnkXeiDJtKmiLpWkmfzTJv5/rSQQcd1Cn9kY98pEwlceWQetCOq7AvBD4DrATeAk4EnpF0QC/PfYOkJZIez9veIulpSQslfT1u/gRwm5mdBhzTm3ydK6dNNtmkU7p///5lKokrhyxq2pcDNwE7m9nnzOxzwM7AzcCPe3nuyUBLcoOkRuBnwJGEPuInSNoV2JYwyAdgdS/zda5s/vjHP5ZMu9qWRdDeE/ixma3JbYj3fwJ8qDcnNrP7gVfzNu8LLDSzZ83sHcKPw8eBRYTADT67oatio0ePXruYb79+/Xw+7TqTRfBaBuxQYPsOwOsp5DeMdTVqCMF6GHA78ElJPwemF3uwpNMlzZM0b+lS75HoKk9raytSmManoaHBZ/qrM1l0+bsZuF7SBcCfCXOPHAT8gNBs0tfyJ6UCMDN7Ezi5qweb2STCfCk0Nzf7ZXlXcZqamhg2bBhtbW0MHTrU59OuM1kE7QsIgfSGRH6rgJ8DXy/2oF5YBGyXSG9LGNTjXE1ob29n8eLFACxevJj29nYP3HUk9eYRM3vHzM4FtiC0b38I2NLMxsc25772MLCTpB0kbQwcD0zryQkkjZU0admyZSkUz7nemTJlCh0dHQB0dHQwderUMpfIZSmzC3JmtsLMHjOzR81sRV+cU9JNwFzgfZIWSfq8mXUAZwEzgSeBW8zsiR6W1VdjdxVr1qxZawfUmBl33XVXmUvkspRK84ikacCJZvZGvF+UmW1wn2kzO6HI9hnAjA09r3OVbPDgwbzwwgud0q5+pNWm3c66xQ58CjLn+tCLL75YMu1qWypB28xOLnS/WkgaC4wdOXJkuYvinHOdZD7IRNKAOLf29lnn3V3epu0q2eGHH94pPXr06DKVxJVDFnOPTJb0xXh/Y+AvhLm1n5Z0ZNr5O1drjjvuuE7pT33qU2UqiSuHLGraRwAPxvvHAO8mLD82Id4qjnf5c5Vs2rTO1/anTy86wNfVoCyC9hbAkni/Bfi1mS0hjJTcNYP8e8ybR1wlmz17dqd0bukxVx+yCNr/AnaLs+8dQVh+DGAgYWSkc64HfMKo+pZF0L4B+BXwOGFK1NxaSfsBT2WQv3M1pbW1lYaG8NVtbGz0CaPqTBbD2C8GTiFMwnRQYuh6B/DDtPN3rtY0NTXR0tKCJFpaWnzekTqTxYRRmNmvC2ybkkXeG8L7abtK19raSltbm9ey61DqQVvSccDrZnZXTH8LOB14AjjJzF5Kuww9ZWbTgenNzc2nlbssrnpdeeWVLFy4MJVz52b5u/jii/v83CNHjuTss8/u8/O6vpFFm/aE3B1JewHfACYCG9H75cacq0srV65k5cqV5S6GK4Msmke2B56O9/8f8Fszu0zSXYSZ+JyrSWnWVs8991wArrjiitTycJUpi5r2W4QBNQCHs67L37LEduecc92QRU37j8CPJT0ANAPHxu0703ktx4rhFyKdc5Uqi5r2WcA7hGB9hpnl5pE8kgptHvERkc65SpV6TdvMFgFjC2z/ctp5O+dcrclkalZJm0g6VtLXJG0et71X0pZZ5O+cc7Uii37aIwkXHwcCmwO3Aq8DZ8b0qWmXwTnnakUWNe2fEubP3hpIdiydBozKIH/nnKsZWfQe+TCwv5mtlpTc/jwwNIP8e8x7jzjnKlVWy41tVGDbcEJf7YrjvUecc5Uqi6B9F/CVRNokbQZ8B/h9Bvk751zNyKJ55CvAvZKeBjYhzK09EngZOK7UA51zznWWRT/tFyXtCZwA7EWo3U8CbjQzn/HGOed6INWgLWkj4BfAN8zsBsIqNs5VlDSnUE1Lrry5iaOqgU/52jdSDdpmtkrSR4EL08yn3BYvXkzjimUMeGpGuYtS0xpXtLN4cUefn3fhwoU888RfGT5wdZ+fOy0brwqXo95+bl6ZS9I9zy9vLHcRakYWbdq3A58ALs8gL+c2yPCBq/nGXm+Uuxg16/vzNyt3EWpGFkH7eeCbkj4CzAPeTO40s59kUIYe6Wk/7WHDhvGvt/ux8v1HpVuwOjfgqRkMG7Z1uYvhXFllEbRPAl4DPhhvSQZUXND25cacc5Uqi94jO6Sdh3PO1YusRkQCIGmgpIFZ5umcc7Ukq6lZvyzpecKw9WWSXpA0XnmTkTjnnCsti6lZLwNOB34EzI2bDwC+BWwDXJB2GZxzrlZkcSHyVOBUM7stse2eOKz9GjxoO+dct2XVpv1okW2Ztqk751y1yyJoTgW+VGD7mcD/ZpC/c87VjCyaR/oDn5F0BPBg3LYfYQGEGyVNzB1oZudkUB7nnKtaWQTt9wPz4/3t499/xdsuieMsg7I451xVy2Jwja8D6ZxzfST1Nm1JI0rs+3Da+W8ISWMlTVq2rCJXQ3PO1bEsLkT+XdKJyQ2SGiRdDNybQf495mtEOucqVRZB+2vA1ZJ+KWkzSe8lDLI5BV/IDRAAAAy6SURBVPhYBvk751zNSD1om9nVQDPwPuBx4K/AIuCDZjY77fydc66WZDW45V9AG7A1MAC408xezShv55yrGVlciDyYMPpxW+ADhGaRH0n6jaSmtPN3zrlakkVNezZhVOSBZrbQzP4X2BMYAjyWQf7OOVczshhcM9rM7k9uMLO2WAP/Rgb5O+dczchicM39RbavAS5JO3/nnKslqTWPSPqzpM0T6UslbZlID44LIzjnnOumNNu09wc2TqS/BGyeSDcSLk4655zrpiznsy60tJhPEuWccz3gixA451wVSTNoG+vXpL1m7ZxzvZBm7xEBv5D0dkxvAlwraUVM908xb+ecq0lpBu0peelfFDhmaor5O+dczUktaJvZyWmde0NJ2hG4CBhkZseWuzzOOddTVXMhUtINkpZIejxve4ukpyUtlPT1Uucws2fN7PPpltQ559KTxTD2vjIZuIpEk4qkRuBnwBjCdK8PS5pG6AN+ad7jTzGzJdkU1Tnn0lE1QdvM7i+wdNm+wEIzexZA0s3Ax83sUuDoLMvXuOJVBjw1o8/P2/DWG2jNqj4/b5qsYSPWbLJZn5+3ccWrhNl9+9bixYt59fV+fOG+Lfr0vKvWiDVV2F+qQbBRQ98W/O3VYst+i/v0nPWqaoJ2EcOAFxLpRcB+xQ6OU8F+D/iQpAtjcM8/5nTgdIDhw4d3qxAjR47sQZF7ZvHiDlauXJna+dMwYMAAhg3r++AKW6fyWm+++ebpvMZvvw1r1vT9edPW0EBD/77t3DWA8Dq73pNZ9VQFYk37DjPbLaY/BRxhZqfG9OeAfc3s7L7Ir7m52ebNm9cXp3LO9YCkR8ysudzlqERVcyGyiEXAdon0tsCLvT2pr8bunKtU1R60HwZ2krSDpI2B44FpvT2pr8bunKtUVRO0Jd1EWMX9fZIWSfq8mXUAZwEzgSeBW8zsiXKW0znn0lQ1FyLN7IQi22cAfd9twznnKlDV1LSz5G3azrlK5UG7AG/Tds5VKg/azjlXRTxoO+dcFfGgXYC3aTvnKpUH7QK8Tds5V6k8aDvnXBXxoO2cc1XEg3YB3qbtnKtUHrQL8DZt51yl8qDtnHNVpKrm086apKXAc+UuR4oGA6+UuxBug9Xy+7e9mQ0pdyEqkQftOiZpnk80X738/atP3jzinHNVxIO2c85VEQ/a9W1SuQvgesXfvzrkbdrOOVdFvKbtnHNVxIN2jZG0PC99kqSr4v0zJI3r4vFrj3flk/8+OpdTNWtEut4zs6vLXQbnXO94TbuOSJog6fx4fx9Jj0qaK+lHkh5PHDpU0p2SnpF0WZmK6/JI2l7S3fF9u1vScEmNkp5VsLmkNZIOjsf/UdLIcpfb9S0P2rVngKS/5W7AxUWO+x/gDDM7AFidt29P4NPA7sCnJW2XXnFdD1wFTDWzDwI3AhPNbDXwD2BX4CDgEeAjkvoD25rZwrKV1qXCg3btWWlme+ZuwLfyD5C0OfBuM/tz3PTLvEPuNrNlZvYWsADYPt0iu246gHXv1f8SgjTAH4GD4+3SuH0f4OGsC+jS50G7PqmL/W8n7q/Gr31Uqlx/3T8CHwH2BWYAmwOHAveXp1guTR6065CZvQb8W9L+cdPx5SyP67Y/s+69+izwQLz/EPBhYE387+hvwBcIwdzVGA/a9evzwCTp/7d3/rFeV2Ucf72vKYUyxiThwnIQ+eOagzGkwO7Qsh8i2WZzZCUEt2ytHyLpNLNNa8LMzGpzmIUJOgaLNqGkWvwIfzSoFhAWxi+7iGKERMAlEYinP57z7Z778Xu/lwvY937ieW2f3e/5nPN5znM+9/t9zvk8n3Oeo1X4yDt2fOhZ9Jb0YnZ8GbgRmCppPTAJmAZgZq8B24HV6dqngT7As3XQO3iDiRWRpyiSzjKztvT5K0CjmU2rs1pBEHRB+CpPXSZIuh3/DmwDptRXnSAIjoUYaQdBEJSI8GkHQRCUiDDaQRAEJSKMdhAEQYkIox1UJcUp+VPXJWvKuFySSep/gnKGJDmxH2JwyhNGu2RImpMM2OwqefemvCe6KdMkXXvytDx+JK2sEhp2O9CILxoJglOaMNrlZDseyOnMyglJb8IXXLxQN63eIMzs32b2NzM7Um9d/tdIapB0Wr31CHoOYbTLyXpgMzAxOzcBOAisLBaWNFXSBkkHJW2SNF1SQ8prTcUWphF3a+Ha6yRtlbRf0qKKq0PSOEmHJQ0slJ+RVuy9DklnS5qfVvi9KunPkqZm+XOAy4AvJF0suUZe5x5J9f82tWmnpO9IOiPLXylplqSZkl6R9HdJ91Xa3RmSxkhaIemApL0pBOqglNdL0ndTfQclrZbUnPIaUru+VJB3ftJ9ZEr3lfSDpM9+SU8W2jVFUpukq5J76hDQJA+l+6vUln2SnpE0tkpdTybdNiYZbZKmZGUGS1ogaU86lkg6r9Y9CXoWYbTLy8NAS5ZuwcOtdph4L+kGYCYe7a8JuBm4Dfh8KjI6/b0Bd0GMzi4fgodovQb4IDASmAFgZk8BW4H/7oSTDOLkpFs13gysAT4MvBP4HvCQpCtS/jRgVWpHYzq2F4VIGgz8AlibdPo08HE8wl3OJ4EjeFyOLwI3pfZURdII4NfAFuA9wBjgx7QvQrs3Xd+S6n0W+KWkRjM7CsxPdRZ12GBmayUJWAIMTvdgJB7UaYWkxsJ9+hoeP+QifPFTHzyyXyUw1Drg51kn2gA8nto7Bl8sdSfQK2tf79S+g3jnOBZ4GViW8oIyYGZxlOgA5gBPAP2AV4HzgIF4ZL5zK/lZ+ReASQUZN+GGpJI24NpCmbvwH3ff7NwdwJYsfQvwXJYen/Q4O6UvT7L712jPAmB2ll4JPFAoMyTJuSSlZ+CGtSErMyXV3TuTs6ogZ2leVxVd5gGrO8k7Ex/1Ts7OnYZ3XHen9PCk5zuyMpuB29Pn9wFtwFsKstcBt2btMGBUF98D4Qb3+pT+EG6wB2dlLk2ypqR0S9JHhTbsBibW+7sdx7EdMdIuKeaR+h7Hf4ifAlaaWQd/tqS3Am/DR7NtlQO4Bxh2DNVsM7M8kNQO4JwsPRd4u6RLU7oFWGRmu6sJk++ycod855XdSZeP4p1Nd2jCDfLR7NwzwBlAvlNL0U1T1L/ISGB5J3nDgNOB31ROmG9AsAofDWNm6/HR9ycAJL07XVeJgT0K6A3sKvw/Lqbj/+MIhZeuks6R9FByb+0F9qe2VO7dhcAOM3spu+z3QH6PRgFD8QiPlbr34gOAY/k+BD2AiD1Sbn6EG842qmx2QLv763N4WM/ucriQtkwmZrZL0k+BFkkbgY8AV9eQdwvunpmGG7c23HVTy5BWQxTcQAUdK9TUvxO5XeVVqzc/Nw/vvL6Bu0aeNrNtKa8B2Im7OIrsyz6/ljqEnLnAAGA60Io/VSzHO6qKfl3FpGjAO4NqoXj/0cW1QQ8hjHa5WY4/svcHFhUzzWynpJeAYWb2aA05h/HH5OPhh8BPgOdxg7SsRtlm4Gdm9hhA8vGeD/wzK3PoGHTZAEyU1JCNtpvTtVu73YJ21uAujGpsSfKb8bYin9WR7yYDbrRnymOVfwz3TefyB+Bxr5/vpm7NwI1mtiTVPQD3+Vd4DhgsaZCZ7UjnLqFjJ7UG9/2/Ymb5PQ9KRLhHSoy5U3I4MNQ8pnI17gJulc8YuUDSxZImyyP8VWgFrpA0UFK/bqqxFPeJ3gk8UnBZFNmU6mmWdCG+5+HQQplW4F1pxkj/TmZ7zAIGAbMkNUmagLt8HjCzf3VT/5xvASPT7I4R6X59RtK5ZnYAeBC4J83KaErpAUkfAMzsRfzl4veBvsDCTP4y3L2yWNJ4SUMljZX0dUnVRt85m4DrJV0kaTT+LuBQlr8U2AjMTbqPAe7HXS2VEfg8vGNdLOmyVP84Sd+OGSTlIYx2yTGz/Wa2r0b+bPxxfRLwRzxA/meBv2bFbgbei8/UWNvN+g2f7XF6+luLu4Hf4TM/ngIO4IYk5z7cGG0AdlHF3538tuNxH/Q63E00H/hqd3SvIncd8H7cP7wa3xHmOtrdLLfhs0keSfUOB640s5cLoh4DRgBL8hFtuldXASvwJ5SNSd4FuL+9Fi3AWfjGvQvwNrdmso/is3x64fd4Lv7C1vAXyqQObRz+pLAQ+Esq1w/Y00X9QQ8hQrMGJ4ykB/EZEx+oty5BO2kK4zp81s0f6q1PcHIIn3Zw3Ejqi89ImEzHhT5BHZB0Df70shmfJnk//nS1po5qBSeZMNrBibAYX+jxcOUFWVBX+gDfxKd57sHnqk+3eJz+vyLcI0EQBCUiXkQGQRCUiDDaQRAEJSKMdhAEQYkIox0EQVAiwmgHQRCUiDDaQRAEJeI/HUJg7v18JkAAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f26de0cccf8>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#quick plot for comparing expression of highly methylated genes versus the rest\n",
    "f, ax = plt.subplots(figsize=(4, 7))\n",
    "ax.set_yscale(\"log\")\n",
    "sns.boxplot(y='UG_01', x=\"Top_500_5mC_gene_body_50\", data=h_contig_df)\n",
    "#title = 'Expression of 500 most highly methylated genes' + '\\n' and the lowly methylated genes'\n",
    "plt.title('Expression of 500 genes with the highest\\ncytosine methylation and lowly methylated genes',fontsize=14)\n",
    "plt.ylabel('Expression in ungerminated spore stage (RPKM)', fontsize=14)\n",
    "plt.xlabel('Methylation coverage', fontsize=14)\n",
    "plt.rc('xtick', labelsize=14)\n",
    "plt.rc('ytick', labelsize=14)\n",
    "plt.savefig('testfig2.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MannwhitneyuResult(statistic=1208273.5, pvalue=3.5252246099440555e-17)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#do some quick stats if those are different from each other\n",
    "mannwhitneyu(h_contig_df[h_contig_df[\"Top_500_5mC_gene_body_50\"] == 'Low']['UG_01'],h_contig_df[h_contig_df[\"Top_500_5mC_gene_body_50\"] == 'High']['UG_01'] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "mwu_dict = {}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to make graph and save out dictionary of mann whitney u\n",
    "def top500_plot(df, col, mod):\n",
    "    f, ax = plt.subplots(figsize=(4, 7))\n",
    "    ax.set_yscale(\"log\")\n",
    "    sns.boxplot(y='UG_01', x=col, data=df)\n",
    "    name = 'Expression of 500 genes with the highest' + '\\n' + mod + '\\n' + 'and all other genes'\n",
    "    plt.title(name, fontsize=14)\n",
    "    plt.ylabel('Expression in ungerminated spore stage (RPKM)', fontsize=14)\n",
    "    plt.xlabel('Methylation coverage', fontsize=14)\n",
    "    plt.rc('xtick', labelsize=14)\n",
    "    plt.rc('ytick', labelsize=14)\n",
    "    name = mod + '.png'\n",
    "    outfn = col + '.png'\n",
    "    outfn = os.path.join(DIRS['FIGURES'], 'expression', outfn)\n",
    "    plt.savefig(outfn, bbox_inches='tight')\n",
    "    mwu_dict[mod] = mannwhitneyu(df[df[col] == 'Low']['UG_01'],df[df[col] == 'High']['UG_01'] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = h_contig_df\n",
    "col = '5mC_gene_body_50'\n",
    "mod = 'gene body 5mC methylation'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Expression of 500 genes with the highest\n",
      "gene body 5mC methylation\n",
      "and all other genes\n"
     ]
    }
   ],
   "source": [
    "name = 'Expression of 500 genes with the highest' + '\\n' + mod + '\\n' + 'and all other genes'\n",
    "print(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/home/anjuni/analysis/figures/expression/5mC_gene_body_50.png'"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "outfn = col + '.png'\n",
    "outfn = os.path.join(DIRS['FIGURES'], 'expression', outfn)\n",
    "outfn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>GS</th>\n",
       "      <th>HE</th>\n",
       "      <th>IT0</th>\n",
       "      <th>IT6</th>\n",
       "      <th>IT9</th>\n",
       "      <th>UG</th>\n",
       "      <th>5mC_gene_body_50</th>\n",
       "      <th>6mA_gene_body_50</th>\n",
       "      <th>5mC_upstream_50</th>\n",
       "      <th>...</th>\n",
       "      <th>Gene_type</th>\n",
       "      <th>Top_500_5mC_gene_body_50</th>\n",
       "      <th>Top_500_5mC_upstream_50</th>\n",
       "      <th>Top_500_5mC_downstream_50</th>\n",
       "      <th>Top_500_5mC_both_50</th>\n",
       "      <th>Top_500_6mA_tss_50</th>\n",
       "      <th>UG_01</th>\n",
       "      <th>HE_01</th>\n",
       "      <th>IT6_01</th>\n",
       "      <th>IT9_01</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_model_hcontig_0000_03.1</td>\n",
       "      <td>0.487427</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.377196</td>\n",
       "      <td>0.384531</td>\n",
       "      <td>0.239374</td>\n",
       "      <td>0.147651</td>\n",
       "      <td>0.213</td>\n",
       "      <td>...</td>\n",
       "      <td>Effector</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.484531</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>2.477196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_model_hcontig_0000_03.2</td>\n",
       "      <td>0.577405</td>\n",
       "      <td>144.208849</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1016.181477</td>\n",
       "      <td>456.791691</td>\n",
       "      <td>0.446980</td>\n",
       "      <td>0.190050</td>\n",
       "      <td>0.110448</td>\n",
       "      <td>0.176</td>\n",
       "      <td>...</td>\n",
       "      <td>Effector</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.546980</td>\n",
       "      <td>144.308849</td>\n",
       "      <td>1016.281477</td>\n",
       "      <td>456.891691</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_model_hcontig_0000_03.3</td>\n",
       "      <td>564.783136</td>\n",
       "      <td>52.086774</td>\n",
       "      <td>24.713948</td>\n",
       "      <td>235.779282</td>\n",
       "      <td>239.133947</td>\n",
       "      <td>316.475087</td>\n",
       "      <td>0.172863</td>\n",
       "      <td>0.085502</td>\n",
       "      <td>0.148</td>\n",
       "      <td>...</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>316.575087</td>\n",
       "      <td>52.186774</td>\n",
       "      <td>235.879282</td>\n",
       "      <td>239.233947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_model_hcontig_0000_03.4</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.167865</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.115347</td>\n",
       "      <td>0.179739</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.224765</td>\n",
       "      <td>0.177230</td>\n",
       "      <td>0.177</td>\n",
       "      <td>...</td>\n",
       "      <td>All genes</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.267865</td>\n",
       "      <td>0.215347</td>\n",
       "      <td>0.279739</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>EVM prediction%2hcontig_0000_003.5</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.127193</td>\n",
       "      <td>0.245614</td>\n",
       "      <td>0.134</td>\n",
       "      <td>...</td>\n",
       "      <td>TE gene</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>High</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "      <td>0.100000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                              gene_ID          GS          HE        IT0  \\\n",
       "0        gene_model_hcontig_0000_03.1    0.487427    0.000000   0.000000   \n",
       "1        gene_model_hcontig_0000_03.2    0.577405  144.208849   0.000000   \n",
       "2        gene_model_hcontig_0000_03.3  564.783136   52.086774  24.713948   \n",
       "3        gene_model_hcontig_0000_03.4    0.000000    0.167865   0.000000   \n",
       "4  EVM prediction%2hcontig_0000_003.5    0.000000    0.000000   0.000000   \n",
       "\n",
       "           IT6         IT9          UG  5mC_gene_body_50  6mA_gene_body_50  \\\n",
       "0     0.000000    2.377196    0.384531          0.239374          0.147651   \n",
       "1  1016.181477  456.791691    0.446980          0.190050          0.110448   \n",
       "2   235.779282  239.133947  316.475087          0.172863          0.085502   \n",
       "3     0.115347    0.179739    0.000000          0.224765          0.177230   \n",
       "4     0.000000    0.000000    0.000000          0.127193          0.245614   \n",
       "\n",
       "   5mC_upstream_50     ...      Gene_type  Top_500_5mC_gene_body_50  \\\n",
       "0            0.213     ...       Effector                      High   \n",
       "1            0.176     ...       Effector                      High   \n",
       "2            0.148     ...      All genes                      High   \n",
       "3            0.177     ...      All genes                      High   \n",
       "4            0.134     ...        TE gene                      High   \n",
       "\n",
       "   Top_500_5mC_upstream_50  Top_500_5mC_downstream_50  Top_500_5mC_both_50  \\\n",
       "0                     High                       High                 High   \n",
       "1                     High                       High                 High   \n",
       "2                     High                       High                 High   \n",
       "3                     High                       High                 High   \n",
       "4                     High                       High                 High   \n",
       "\n",
       "   Top_500_6mA_tss_50       UG_01       HE_01       IT6_01      IT9_01  \n",
       "0                High    0.484531    0.100000     0.100000    2.477196  \n",
       "1                High    0.546980  144.308849  1016.281477  456.891691  \n",
       "2                High  316.575087   52.186774   235.879282  239.233947  \n",
       "3                High    0.100000    0.267865     0.215347    0.279739  \n",
       "4                High    0.100000    0.100000     0.100000    0.100000  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h_contig_df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run the plots for each type: 5mc up, down, both, gene body, 6ma tss\n",
    "top500_plot(h_contig_df, '5mC_gene_body_50', 'gene body 5mC methylation')\n",
    "top500_plot(h_contig_df, '5mC_upstream_50', 'upstream 5mC methylation')\n",
    "top500_plot(h_contig_df, '5mC_downstream_50', 'downstream 5mC methylation')\n",
    "top500_plot(h_contig_df, '5mC_both_50', 'upstream and downstream 5mC methylation')\n",
    "top500_plot(h_contig_df, '6mA_tss_50', 'transcription start site 6mA methylation')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Gene types"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genetypelog_plot(df, col, mod):\n",
    "    f, ax = plt.subplots(figsize=(4, 7))\n",
    "    ax.set_yscale(\"log\")\n",
    "    sns.boxplot(y=col, x='Gene_type', data=df)\n",
    "    title = mod + '\\n' + 'of gene types'\n",
    "    plt.title(title,fontsize=14)\n",
    "    plt.ylabel('Methylation coverage level', fontsize=14)\n",
    "    plt.xlabel('Gene type', fontsize=14)\n",
    "    plt.rc('xtick', labelsize=14)\n",
    "    plt.rc('ytick', labelsize=14)\n",
    "    plt.savefig('testfig2.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genetype_plot(df, col, mod):\n",
    "    f, ax = plt.subplots(figsize=(4, 7))\n",
    "    sns.boxplot(y=col, x='Gene_type', data=df)\n",
    "    title = mod + '\\n' + 'of gene types'\n",
    "    plt.title(title,fontsize=14)\n",
    "    plt.ylabel('Methylation coverage level', fontsize=14)\n",
    "    plt.xlabel('Gene type', fontsize=14)\n",
    "    plt.rc('xtick', labelsize=14)\n",
    "    plt.rc('ytick', labelsize=14)\n",
    "    plt.savefig('testfig2.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, ax = plt.subplots(figsize=(7, 6))\n",
    "ax.set_yscale(\"log\")\n",
    "sns.boxplot(y='5mC_gene_body_50', x='Gene_type', data=h_contig_df)\n",
    "plt.title('5mC Gene body methylation of gene types',fontsize=14)\n",
    "plt.xlabel('Gene type', fontsize=14)\n",
    "plt.ylabel('Methylation coverage level', fontsize=14)\n",
    "plt.rc('xtick', labelsize=14) \n",
    "plt.rc('ytick', labelsize=14) \n",
    "plt.savefig('testfig3.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f, ax = plt.subplots(figsize=(6, 6))\n",
    "ax.set_yscale(\"log\")\n",
    "sns.boxplot(y='UG_01', x='Gene_type', data=h_contig_df)\n",
    "plt.title('Gene expression in ungerminated spore stage of gene types',fontsize=14)\n",
    "plt.xlabel('Gene type', fontsize=14)\n",
    "plt.ylabel('Gene expression during ungerminated spore stage (RPKM)', fontsize=14)\n",
    "plt.rc('xtick', labelsize=14) \n",
    "plt.rc('ytick', labelsize=14) \n",
    "plt.savefig('testfig4.png', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "h_contig_df.loc[h_contig_df.Protein_ID.isin(effctor_list), 'Effector'] = 'True'\n",
    "f, ax = plt.subplots(figsize=(7, 6))\n",
    "ax.set_xscale(\"log\")\n",
    "sns.boxplot(x='HE_01', y='Gene_type', data=h_contig_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in ['5mC_gene_body_50',\n",
    "       '6mA_gene_body_50', '5mC_upstream_50', '6mA_upstream_50',\n",
    "       '5mC_upstream_50.1', '6mA_upstream_50.1', '5mC_downstream_50',\n",
    "       '6mA_downstream_50', '5mC_gene_body_95', '6mA_gene_body_95',\n",
    "       '5mC_upstream_95', '6mA_upstream_95', '5mC_downstream_95',\n",
    "       '6mA_downstream_95']:\n",
    "    print(x)\n",
    "    f, ax = plt.subplots(figsize=(7, 6))\n",
    "    #ax.set_xscale(\"log\")\n",
    "    sns.boxplot(x=x, y='Gene_type', data=h_contig_df)\n",
    "    plt.title(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for x in [ '5mC_gene_body_95', '6mA_gene_body_95',\n",
    "       '5mC_upstream_95', '6mA_upstream_95', '5mC_downstream_95',\n",
    "       '6mA_downstream_95']:\n",
    "    print(x)\n",
    "    f, ax = plt.subplots(figsize=(7, 6))\n",
    "    #ax.set_xscale(\"log\")\n",
    "    ax.set_xlim(-0.1,0.1)\n",
    "    sns.boxplot(x=x, y='Gene_type', data=h_contig_df)\n",
    "    plt.title(x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
